// @deepkit/mongo/dist/cjs/benchmarks/bench-new-client.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
const mongodb_1 = require("mongodb");
const client_1 = require("../src/client/client");
const ismaster_1 = require("../src/client/command/ismaster");
const find_1 = require("../src/client/command/find");
const empty_1 = require("../src/client/command/empty");
const type_1 = require("@deepkit/type");
const delete_1 = require("../src/client/command/delete");
const insert_1 = require("../src/client/command/insert");
const utils_1 = require("./utils");
let User = class User {
    constructor(id, name) {
        this.id = id;
        this.name = name;
        this.tags = [];
        this.priority = 0;
    }
};
__decorate([
    type_1.f.mongoId.primary,
    __metadata("design:type", String)
], User.prototype, "_id", void 0);
__decorate([
    type_1.f,
    __metadata("design:type", Boolean)
], User.prototype, "ready", void 0);
__decorate([
    type_1.f.array(type_1.f.string),
    __metadata("design:type", Array)
], User.prototype, "tags", void 0);
__decorate([
    type_1.f,
    __metadata("design:type", Number)
], User.prototype, "priority", void 0);
User = __decorate([
    type_1.Entity('user'),
    __param(0, type_1.f),
    __param(1, type_1.f),
    __metadata("design:paramtypes", [Number, String])
], User);
exports.User = User;
async function main() {
    const items = [];
    const count = 10000;
    for (let i = 1; i <= count; i++) {
        const user = new User(i, 'Peter ' + i);
        user.ready = true;
        user.priority = 5;
        user.tags = ['a', 'b', 'c'];
        items.push(user);
    }
    const mongoClient = await mongodb_1.MongoClient.connect(`mongodb://127.0.0.1/benchmark-a`, {
        useUnifiedTopology: true,
        useNewUrlParser: true,
    });
    const collection = mongoClient.db('benchmark-a').collection('user');
    const client = new client_1.MongoClient('mongodb://127.0.0.1/benchmark-a');
    await client.connect();
    await collection.deleteMany({});
    await utils_1.bench(1, 'native mongodb insert 10k', async () => {
        await collection.insertMany(items);
    });
    for (const item of items) {
        item._id = undefined;
    }
    await client.execute(new delete_1.DeleteCommand(User, {}));
    await utils_1.bench(1, 'new client: insert 10k', async () => {
        await client.execute(new insert_1.InsertCommand(User, items));
    });
    await utils_1.bench(2000, 'reference', async () => {
        await new Promise((resolve) => {
            resolve(undefined);
        });
    });
    await utils_1.bench(10000, 'new client: EmptyCommand', async () => {
        const response = await client.execute(new empty_1.EmptyCommand);
    });
    await utils_1.bench(10000, 'new client: IsMasterCommand', async () => {
        const response = await client.execute(new ismaster_1.IsMasterCommand);
    });
    await utils_1.bench(10000, 'ori Mongodb find 1', async () => {
        const item = await collection.find({}).limit(1).toArray();
    });
    await utils_1.bench(10000, 'new client: FindCommand 1', async () => {
        const cmd = new find_1.FindCommand(User);
        cmd.limit = 1;
        const response = await client.execute(cmd);
    });
    await utils_1.bench(10000, 'ori Mongodb find 10', async () => {
        const item = await collection.find({}).limit(10).toArray();
    });
    await utils_1.bench(10000, 'new client: FindCommand 10', async () => {
        const cmd = new find_1.FindCommand(User);
        cmd.limit = 10;
        const response = await client.execute(cmd);
    });
    {
        const items = await collection.find({}).toArray();
        if (items.length !== 10000)
            throw new Error(`Invalid, got ${items.length}`);
    }
    await utils_1.bench(100, 'ori Mongodb find 10k', async () => {
        const item = await collection.find({}).batchSize(10001).toArray();
    });
    {
        const items = await client.execute(new find_1.FindCommand(User));
        if (items.length !== 10000)
            throw new Error(`Invalid, got ${items.length}`);
    }
    await utils_1.bench(100, 'new client: FindCommand 10k', async () => {
        const response = await client.execute(new find_1.FindCommand(User));
    });
    client.close();
    mongoClient.close();
}
main();
