// @deepkit/mongo/dist/cjs/src/client/command/aggregate.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregateCommand = void 0;
const core_1 = require("@deepkit/core");
const type_1 = require("@deepkit/type");
const command_1 = require("./command");
const aggregateSchema = type_1.t.schema({
    aggregate: type_1.t.string,
    $db: type_1.t.string,
    pipeline: type_1.t.array(type_1.t.any),
    cursor: {
        batchSize: type_1.t.number,
    }
});
class AggregateCommand extends command_1.Command {
    constructor(classSchema, pipeline = [], resultSchema) {
        super();
        this.classSchema = classSchema;
        this.pipeline = pipeline;
        this.resultSchema = resultSchema;
        this.partial = false;
    }
    async execute(config) {
        const schema = type_1.getClassSchema(this.classSchema);
        const cmd = {
            aggregate: schema.collectionName || schema.name || 'unknown',
            $db: schema.databaseSchemaName || config.defaultDb || 'admin',
            pipeline: this.pipeline,
            cursor: {
                batchSize: 20000,
            }
        };
        const resultSchema = this.resultSchema || schema;
        const jit = resultSchema.jit;
        let specialisedResponse = this.partial ? jit.mdbAggregatePartial : jit.mdbAggregate;
        if (!specialisedResponse) {
            if (this.partial) {
                specialisedResponse = type_1.t.extendSchema(command_1.BaseResponse, {
                    cursor: {
                        id: type_1.t.number,
                        firstBatch: type_1.t.array(type_1.t.partial(resultSchema)),
                        nextBatch: type_1.t.array(type_1.t.partial(resultSchema)),
                    },
                });
                jit.mdbAggregatePartial = specialisedResponse;
            }
            else {
                specialisedResponse = type_1.t.extendSchema(command_1.BaseResponse, {
                    cursor: {
                        id: type_1.t.number,
                        firstBatch: type_1.t.array(resultSchema),
                        nextBatch: type_1.t.array(resultSchema),
                    },
                });
                jit.mdbAggregate = specialisedResponse;
            }
            core_1.toFastProperties(jit);
        }
        const res = await this.sendAndWait(aggregateSchema, cmd, specialisedResponse);
        //todo: implement fetchMore
        return res.cursor.firstBatch;
    }
    needsWritableHost() {
        return false;
    }
}
exports.AggregateCommand = AggregateCommand;
