// @deepkit/mongo/dist/cjs/src/client/command/count.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountCommand = void 0;
const command_1 = require("./command");
const type_1 = require("@deepkit/type");
class CountResponse extends type_1.t.extendClass(command_1.BaseResponse, {
    n: type_1.t.number,
}) {
}
const countSchema = type_1.t.schema({
    count: type_1.t.string,
    $db: type_1.t.string,
    limit: type_1.t.number.optional,
    query: type_1.t.any,
    skip: type_1.t.number.optional,
});
class CountCommand extends command_1.Command {
    constructor(classSchema, query = {}, limit = 0, skip = 0) {
        super();
        this.classSchema = classSchema;
        this.query = query;
        this.limit = limit;
        this.skip = skip;
    }
    async execute(config) {
        const schema = type_1.getClassSchema(this.classSchema);
        const cmd = {
            count: schema.collectionName || schema.name || 'unknown',
            $db: schema.databaseSchemaName || config.defaultDb || 'admin',
            query: this.query,
            limit: this.limit,
            skip: this.skip,
        };
        const res = await this.sendAndWait(countSchema, cmd, CountResponse);
        return res.n;
    }
    needsWritableHost() {
        return false;
    }
}
exports.CountCommand = CountCommand;
