// @deepkit/mongo/dist/cjs/src/client/command/delete.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteCommand = void 0;
const command_1 = require("./command");
const type_1 = require("@deepkit/type");
class DeleteResponse extends type_1.t.extendClass(command_1.BaseResponse, {
    n: type_1.t.number,
}) {
}
const deleteSchema = type_1.t.schema({
    delete: type_1.t.string,
    $db: type_1.t.string,
    deletes: type_1.t.array({
        q: type_1.t.any,
        limit: type_1.t.number,
    }),
});
class DeleteCommand extends command_1.Command {
    constructor(classSchema, filter = {}, limit = 0, skip = 0) {
        super();
        this.classSchema = classSchema;
        this.filter = filter;
        this.limit = limit;
        this.skip = skip;
    }
    async execute(config) {
        const schema = type_1.getClassSchema(this.classSchema);
        const cmd = {
            delete: schema.collectionName || schema.name || 'unknown',
            $db: schema.databaseSchemaName || config.defaultDb || 'admin',
            deletes: [
                {
                    q: this.filter,
                    limit: this.limit,
                }
            ]
        };
        const res = await this.sendAndWait(deleteSchema, cmd, DeleteResponse);
        return res.n;
    }
    needsWritableHost() {
        return false;
    }
}
exports.DeleteCommand = DeleteCommand;
