// @deepkit/mongo/dist/cjs/src/client/command/find.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FindCommand = void 0;
const command_1 = require("./command");
const type_1 = require("@deepkit/type");
const core_1 = require("@deepkit/core");
const findSchema = type_1.t.schema({
    find: type_1.t.string,
    $db: type_1.t.string,
    batchSize: type_1.t.number,
    limit: type_1.t.number,
    skip: type_1.t.number,
    filter: type_1.t.any,
    projection: type_1.t.any.optional,
    sort: type_1.t.any.optional,
});
class FindCommand extends command_1.Command {
    constructor(classSchema, filter = {}, projection, sort, limit = 0, skip = 0) {
        super();
        this.classSchema = classSchema;
        this.filter = filter;
        this.projection = projection;
        this.sort = sort;
        this.limit = limit;
        this.skip = skip;
    }
    async execute(config) {
        const schema = type_1.getClassSchema(this.classSchema);
        const cmd = {
            find: schema.collectionName || schema.name || 'unknown',
            $db: schema.databaseSchemaName || config.defaultDb || 'admin',
            filter: this.filter,
            limit: this.limit,
            skip: this.skip,
            batchSize: 1000000, //todo make configurable
        };
        if (this.projection)
            cmd.projection = this.projection;
        if (this.sort)
            cmd.sort = this.sort;
        const jit = schema.jit;
        let specialisedResponse = this.projection ? jit.mdbFindPartial : jit.mdbFind;
        if (!specialisedResponse) {
            if (this.projection) {
                specialisedResponse = type_1.t.extendSchema(command_1.BaseResponse, {
                    cursor: {
                        id: type_1.t.number,
                        firstBatch: type_1.t.array(type_1.t.partial(schema)),
                        nextBatch: type_1.t.array(type_1.t.partial(schema)),
                    },
                });
                jit.mdbFindPartial = specialisedResponse;
            }
            else {
                specialisedResponse = type_1.t.extendSchema(command_1.BaseResponse, {
                    cursor: {
                        id: type_1.t.number,
                        firstBatch: type_1.t.array(schema),
                        nextBatch: type_1.t.array(schema),
                    },
                });
                jit.mdbFind = specialisedResponse;
            }
            core_1.toFastProperties(jit);
        }
        const res = await this.sendAndWait(findSchema, cmd, specialisedResponse);
        //todo: implement fetchMore and decrease batchSize
        return res.cursor.firstBatch;
    }
    needsWritableHost() {
        return false;
    }
}
exports.FindCommand = FindCommand;
