// @deepkit/mongo/dist/cjs/src/client/command/find-and-modify.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FindAndModifyCommand = void 0;
const command_1 = require("./command");
const type_1 = require("@deepkit/type");
class FindAndModifyResponse extends type_1.t.extendClass(command_1.BaseResponse, {
    value: type_1.t.any,
}) {
}
const findAndModifySchema = type_1.t.schema({
    findAndModify: type_1.t.string,
    $db: type_1.t.string,
    query: type_1.t.any,
    update: type_1.t.any,
    new: type_1.t.boolean,
    upsert: type_1.t.boolean,
    fields: type_1.t.map(type_1.t.number),
});
class FindAndModifyCommand extends command_1.Command {
    constructor(classSchema, query, update) {
        super();
        this.classSchema = classSchema;
        this.query = query;
        this.update = update;
        this.upsert = false;
        this.fields = [];
        this.returnNew = false;
    }
    async execute(config) {
        const schema = type_1.getClassSchema(this.classSchema);
        const fields = {};
        for (const name of this.fields)
            fields[name] = 1;
        const cmd = {
            findAndModify: schema.collectionName || schema.name || 'unknown',
            $db: schema.databaseSchemaName || config.defaultDb || 'admin',
            query: this.query,
            update: this.update,
            new: this.returnNew,
            upsert: this.upsert,
            fields: fields,
        };
        return await this.sendAndWait(findAndModifySchema, cmd, FindAndModifyResponse);
    }
    needsWritableHost() {
        return false;
    }
}
exports.FindAndModifyCommand = FindAndModifyCommand;
