// @deepkit/mongo/dist/cjs/src/client/command/ismaster.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsMasterCommand = exports.IsMasterResponse = void 0;
//see https://docs.mongodb.com/manual/reference/command/isMaster/
//we add only fields we really need to increase parsing time.
const type_1 = require("@deepkit/type");
const command_1 = require("./command");
class IsMasterResponse extends type_1.t.extendClass(command_1.BaseResponse, {
    ismaster: type_1.t.boolean,
    maxBsonObjectSize: type_1.t.number,
    maxMessageSizeBytes: type_1.t.number,
    maxWriteBatchSize: type_1.t.number,
    minWireVersion: type_1.t.number,
    maxWireVersion: type_1.t.number,
    //indicates that the mongod or mongos is running in read-only mode
    readOnly: type_1.t.boolean,
    compression: type_1.t.array(type_1.t.string),
    saslSupportedMechs: type_1.t.array(type_1.t.string),
    //mongos instances add the following field to the isMaster response document:
    msg: type_1.t.string,
    //isMaster contains these fields when returned by a member of a replica set:
    // hosts: t.array(t.string),
    setName: type_1.t.string,
    // setVersion: t.number, //replica set version
    secondary: type_1.t.boolean,
    arbiterOnly: type_1.t.boolean,
    hidden: type_1.t.boolean,
}) {
}
exports.IsMasterResponse = IsMasterResponse;
const isMasterSchema = type_1.t.schema({
    isMaster: type_1.t.number,
    $db: type_1.t.string,
});
class IsMasterCommand extends command_1.Command {
    needsWritableHost() {
        return false;
    }
    async execute(config, host) {
        const cmd = {
            isMaster: 1,
            $db: config.getAuthSource(),
        };
        return this.sendAndWait(isMasterSchema, cmd, IsMasterResponse);
    }
}
exports.IsMasterCommand = IsMasterCommand;
