// @deepkit/orm/dist/cjs/tests/identity-map.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
require("reflect-metadata");
const type_1 = require("@deepkit/type");
const index_1 = require("../index");
const reference_1 = require("../src/reference");
globals_1.test('getNormalizedPrimaryKey', () => {
    class User {
        constructor() {
            this.id = '';
            this.name = 'Foo';
        }
    }
    __decorate([
        type_1.t.primary,
        __metadata("design:type", String)
    ], User.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], User.prototype, "name", void 0);
    globals_1.expect(index_1.getNormalizedPrimaryKey(type_1.getClassSchema(User), '123')).toEqual({ id: '123' });
    globals_1.expect(index_1.getNormalizedPrimaryKey(type_1.getClassSchema(User), { id: '124' })).toEqual({ id: '124' });
    class User2 {
        constructor() {
            this.id = '';
            this.id2 = '';
            this.name = 'Foo';
        }
    }
    __decorate([
        type_1.t.primary,
        __metadata("design:type", String)
    ], User2.prototype, "id", void 0);
    __decorate([
        type_1.t.primary,
        __metadata("design:type", String)
    ], User2.prototype, "id2", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], User2.prototype, "name", void 0);
    globals_1.expect(() => index_1.getNormalizedPrimaryKey(type_1.getClassSchema(User2), '123')).toThrow('Entity User2 has composite primary key');
    globals_1.expect(index_1.getNormalizedPrimaryKey(type_1.getClassSchema(User2), { id: '124', id2: '444' })).toEqual({ id: '124', id2: '444' });
});
globals_1.test('original schema', () => {
    let User = class User {
        constructor(username) {
            this.username = username;
        }
    };
    __decorate([
        type_1.t.primary,
        __metadata("design:type", Number)
    ], User.prototype, "id", void 0);
    User = __decorate([
        __param(0, type_1.t),
        __metadata("design:paramtypes", [String])
    ], User);
    const ref = reference_1.getReference(type_1.getClassSchema(User), { id: 2 });
    globals_1.expect(ref.id).toBe(2);
    globals_1.expect(type_1.getClassSchema(ref) === type_1.getClassSchema(User)).toBe(true);
});
globals_1.test('snapshot correct state', () => {
    const image = type_1.t.schema({
        id: type_1.t.number.primary
    });
    const user = type_1.t.schema({
        id: type_1.t.number.primary,
        username: type_1.t.string,
        image: type_1.t.type(image).optional.reference(),
        image2: type_1.t.type(image).optional.reference(),
    });
    const formatter = new index_1.Formatter(user, type_1.jsonSerializer);
    {
        const query = new index_1.BaseQuery(user);
        const user1 = formatter.hydrate(query.model, { username: 'Peter', id: '2', image: '1' });
        const snapshot = index_1.getInstanceStateFromItem(user1).getSnapshot();
        globals_1.expect(snapshot.hasOwnProperty('image2')).toBe(true);
        //when schema changes we get from mongodb `undefined` for new fields, but our snapshot converts that to `null`
        // since all databases use `null` as `not defined`. this means we basically ignore `undefined` where possible.
        globals_1.expect(snapshot).toEqual({ username: 'Peter', id: 2, image: { id: 1 }, image2: null });
        user1.image2 = reference_1.getReference(image, { id: 2 });
        globals_1.expect(user1.image2.id).toBe(2);
        index_1.getInstanceStateFromItem(user1).markAsPersisted();
        globals_1.expect(index_1.getInstanceStateFromItem(user1).getSnapshot()).toEqual({ username: 'Peter', id: 2, image: { id: 1 }, image2: { id: 2 } });
        user1.image = undefined;
        index_1.getInstanceStateFromItem(user1).markAsPersisted();
        globals_1.expect(index_1.getInstanceStateFromItem(user1).getSnapshot()).toEqual({ username: 'Peter', id: 2, image: null, image2: { id: 2 } });
    }
    {
        const query = new index_1.BaseQuery(user);
        const user1 = formatter.hydrate(query.model, { username: 'Peter2', id: '3', image: '1', image2: null });
        const snapshot = index_1.getInstanceStateFromItem(user1).getSnapshot();
        globals_1.expect(snapshot.hasOwnProperty('image2')).toBe(true);
        globals_1.expect(snapshot).toEqual({ username: 'Peter2', id: 3, image: { id: 1 }, image2: null });
        user1.image2 = reference_1.getReference(image, { id: 2 });
        globals_1.expect(user1.image2.id).toBe(2);
        index_1.getInstanceStateFromItem(user1).markAsPersisted();
        globals_1.expect(index_1.getInstanceStateFromItem(user1).getSnapshot()).toEqual({ username: 'Peter2', id: 3, image: { id: 1 }, image2: { id: 2 } });
    }
});
