// @deepkit/mongo/dist/esm/src/client/command/delete.js
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
import { BaseResponse, Command } from './command';
import { getClassSchema, t } from '@deepkit/type';
class DeleteResponse extends t.extendClass(BaseResponse, {
    n: t.number,
}) {
}
const deleteSchema = t.schema({
    delete: t.string,
    $db: t.string,
    deletes: t.array({
        q: t.any,
        limit: t.number,
    }),
});
export class DeleteCommand extends Command {
    constructor(classSchema, filter = {}, limit = 0, skip = 0) {
        super();
        this.classSchema = classSchema;
        this.filter = filter;
        this.limit = limit;
        this.skip = skip;
    }
    async execute(config) {
        const schema = getClassSchema(this.classSchema);
        const cmd = {
            delete: schema.collectionName || schema.name || 'unknown',
            $db: schema.databaseSchemaName || config.defaultDb || 'admin',
            deletes: [
                {
                    q: this.filter,
                    limit: this.limit,
                }
            ]
        };
        const res = await this.sendAndWait(deleteSchema, cmd, DeleteResponse);
        return res.n;
    }
    needsWritableHost() {
        return false;
    }
}
