// @deepkit/template/dist/cjs/tests/simple.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simpleHtmlInjectedValid = exports.simpleHtmlInjected = exports.simple5 = exports.spread = exports.simple4 = exports.simple3 = exports.simple2 = exports.simple1 = exports.simpleOnlyChildren = exports.simpleOnlyAttributes = void 0;
const __1 = require("..");
const jsx_runtime_1 = require("../jsx-runtime");
const template_1 = require("../src/template");
function simpleOnlyAttributes() {
    return jsx_runtime_1.jsx("div", { id: "123" }, void 0);
}
exports.simpleOnlyAttributes = simpleOnlyAttributes;
function simpleOnlyChildren() {
    return jsx_runtime_1.jsx("div", { children: "Test" }, void 0);
}
exports.simpleOnlyChildren = simpleOnlyChildren;
function simple1() {
    return jsx_runtime_1.jsx("div", Object.assign({ id: "123" }, { children: "Test" }), void 0);
}
exports.simple1 = simple1;
function simple2() {
    return jsx_runtime_1.jsx("div", Object.assign({ id: "123" }, { children: jsx_runtime_1.jsx("b", { children: "strong" }, void 0) }), void 0);
}
exports.simple2 = simple2;
function simple3() {
    return jsx_runtime_1.jsxs("div", Object.assign({ id: "123" }, { children: [jsx_runtime_1.jsx("b", { children: "strong" }, void 0), jsx_runtime_1.jsx("b", { children: "strong2" }, void 0)] }), void 0);
}
exports.simple3 = simple3;
function simple4() {
    return jsx_runtime_1.jsx("div", Object.assign({ id: "123", class: "active" }, { children: jsx_runtime_1.jsxs("div", { children: [jsx_runtime_1.jsx("b", { children: "strong" }, void 0), jsx_runtime_1.jsx("b", { children: "strong2" }, void 0)] }, void 0) }), void 0);
}
exports.simple4 = simple4;
function spread() {
    const props = { foo: true, bar: false };
    return __1.createElement("div", { ...props, key: "myKey" },
        jsx_runtime_1.jsx("b", { children: "TestBold" }, void 0),
        "Test");
}
exports.spread = spread;
function Website(props) {
    return jsx_runtime_1.jsxs("html", { children: [jsx_runtime_1.jsx("header", { children: jsx_runtime_1.jsx("meta", Object.assign({ name: "title" }, { children: "My Website" }), void 0) }, void 0), jsx_runtime_1.jsx("body", { children: jsx_runtime_1.jsx("div", { children: props.children }, void 0) }, void 0)] }, void 0);
}
function simple5(name) {
    return jsx_runtime_1.jsxs(Website, { children: [jsx_runtime_1.jsx("h1", { children: "Hi" }, void 0), jsx_runtime_1.jsxs("p", { children: ["How are you, ", name, "?"] }, void 0)] }, void 0);
}
exports.simple5 = simple5;
function simpleHtmlInjected() {
    const myHtml = '<strong>MyHTML</strong>';
    return jsx_runtime_1.jsx("div", { children: myHtml }, void 0);
}
exports.simpleHtmlInjected = simpleHtmlInjected;
function simpleHtmlInjectedValid() {
    const myHtml = '<strong>MyHTML</strong>';
    return jsx_runtime_1.jsx("div", { children: template_1.html(myHtml) }, void 0);
}
exports.simpleHtmlInjectedValid = simpleHtmlInjectedValid;
