// @deepkit/injector/dist/cjs/src/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InjectorModule = void 0;
const injector_1 = require("./injector");
class InjectorModule {
    constructor(name, config) {
        this.name = name;
        this.config = config;
        this.setupProviderRegistry = new injector_1.ConfiguredProviderRegistry;
    }
    getName() {
        return this.name;
    }
    getConfig() {
        return this.config;
    }
    setConfig(config) {
        Object.assign(this.config, config);
    }
    getConfiguredProviderRegistry() {
        return this.setupProviderRegistry;
    }
    /**
     * Returns a configuration object that reflects the API of the given ClassType or token. Each call
     * is scheduled and executed once the provider has been created by the dependency injection container.
     */
    setupProvider(classTypeOrToken) {
        return injector_1.setupProvider(classTypeOrToken, this.setupProviderRegistry);
    }
}
exports.InjectorModule = InjectorModule;
