// @deepkit/framework/dist/cjs/tests/broker.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const broker_1 = require("@deepkit/broker");
const core_1 = require("@deepkit/core");
const type_1 = require("@deepkit/type");
const globals_1 = require("@jest/globals");
require("reflect-metadata");
const rxjs_1 = require("rxjs");
const broker_2 = require("../src/broker/broker");
globals_1.test('entity channel number', async () => {
    const kernel = new broker_1.BrokerKernel();
    const client = new broker_2.DirectBroker(kernel);
    let Model = class Model {
        constructor() {
            this.id = 0;
            this.version = 0;
            this.title = '';
        }
    };
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Model.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Model.prototype, "version", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], Model.prototype, "title", void 0);
    Model = __decorate([
        type_1.entity.name('model')
    ], Model);
    {
        const subject = new rxjs_1.BehaviorSubject(undefined);
        const channel = client.entityChannel(Model);
        await channel.subscribe(v => subject.next(v));
        await channel.publishRemove([23]);
        await core_1.sleep(0);
        globals_1.expect(subject.value).toEqual({ type: broker_2.EntityChannelMessageType.remove, ids: [23] });
        await channel.publishPatch(23, 5, { $set: { username: true } }, { title: 'asd' });
        await core_1.sleep(0);
        globals_1.expect(subject.value).toEqual({
            type: broker_2.EntityChannelMessageType.patch,
            id: 23,
            version: 5,
            patch: { $set: { username: true } },
            item: { title: 'asd' }
        });
        await channel.publishAdd(type_1.plainToClass(Model, { id: 1243, version: 0, title: 'peter' }));
        await core_1.sleep(0);
        globals_1.expect(subject.value).toEqual({
            type: broker_2.EntityChannelMessageType.add,
            id: 1243,
            item: { id: 1243, version: 0, title: 'peter' }
        });
    }
});
globals_1.test('entity channel uuid', async () => {
    const kernel = new broker_1.BrokerKernel();
    const client = new broker_2.DirectBroker(kernel);
    let Model = class Model {
        constructor() {
            this.id = type_1.uuid();
            this.version = 0;
            this.title = '';
        }
    };
    __decorate([
        type_1.t.primary.uuid,
        __metadata("design:type", String)
    ], Model.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Model.prototype, "version", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], Model.prototype, "title", void 0);
    Model = __decorate([
        type_1.entity.name('modelUuid')
    ], Model);
    {
        const subject = new rxjs_1.BehaviorSubject(undefined);
        const channel = client.entityChannel(Model);
        await channel.subscribe(v => subject.next(v));
        const item = new Model();
        await channel.publishRemove([item.id]);
        await core_1.sleep(0);
        globals_1.expect(subject.value).toEqual({ type: broker_2.EntityChannelMessageType.remove, ids: [item.id] });
        await channel.publishPatch(item.id, 5, { $set: { username: true } }, { title: 'asd' });
        await core_1.sleep(0);
        globals_1.expect(subject.value).toEqual({
            type: broker_2.EntityChannelMessageType.patch,
            id: item.id,
            version: 5,
            patch: { $set: { username: true } },
            item: { title: 'asd' }
        });
        await channel.publishAdd(type_1.plainToClass(Model, { id: item.id, version: 0, title: 'peter' }));
        await core_1.sleep(0);
        globals_1.expect(subject.value).toEqual({
            type: broker_2.EntityChannelMessageType.add,
            id: item.id,
            item: { id: item.id, version: 0, title: 'peter' }
        });
    }
});
