// @deepkit/framework/dist/cjs/tests/live-database.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@deepkit/core");
const core_rxjs_1 = require("@deepkit/core-rxjs");
const orm_1 = require("@deepkit/orm");
const rpc_1 = require("@deepkit/rpc");
const type_1 = require("@deepkit/type");
const globals_1 = require("@jest/globals");
const assert_1 = require("assert");
require("reflect-metadata");
const rxjs_1 = require("rxjs");
const broker_1 = require("../src/broker/broker");
const live_database_1 = require("../src/database/live-database");
const testing_1 = require("../src/testing");
global['createTestingApp'] || (global['createTestingApp'] = testing_1.createTestingApp);
let UserBase = class UserBase {
    constructor(name) {
        this.name = name;
        this.id = type_1.uuid();
        this.version = 1;
    }
};
__decorate([
    type_1.t.primary.uuid,
    __metadata("design:type", String)
], UserBase.prototype, "id", void 0);
__decorate([
    type_1.t,
    __metadata("design:type", Number)
], UserBase.prototype, "version", void 0);
UserBase = __decorate([
    __param(0, type_1.t),
    __metadata("design:paramtypes", [String])
], UserBase);
globals_1.test('test broker', async () => {
    let User = class User extends UserBase {
    };
    User = __decorate([
        type_1.Entity('user_broker_test')
    ], User);
    const testing = createTestingApp({});
    await testing.startServer();
    const broker = testing.app.get(broker_1.Broker);
    globals_1.expect(broker).toBeInstanceOf(broker_1.DirectBroker);
    const subject = new rxjs_1.BehaviorSubject(undefined);
    await broker.entityChannel(User).subscribe(v => {
        subject.next(v);
    });
    await broker.entityChannel(User).publishPatch(1, 1, { $set: { username: true } }, {});
    await core_1.sleep(0);
    globals_1.expect(subject.value).not.toBe(undefined);
    globals_1.expect(subject.value.id).toBe(1);
    globals_1.expect(subject.value.patch).toEqual({ $set: { username: true } });
});
globals_1.test('test increase', async () => {
    let User = class User extends UserBase {
        constructor() {
            super(...arguments);
            this.connections = 0;
        }
    };
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], User.prototype, "connections", void 0);
    User = __decorate([
        type_1.Entity('user_increase')
    ], User);
    let TestController = class TestController {
        constructor(liveDatabase, database) {
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(User);
        }
        async start() {
            await this.database.query(User).deleteMany();
            const user = new User('peter');
            await this.database.persist(user);
        }
        async increase(i) {
            await this.database.query(User).patchOne({ $inc: { connections: i } });
        }
        async user() {
            return await this.liveDatabase.query(User).findOne();
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "start", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "increase", null);
    __decorate([
        rpc_1.rpc.action(),
        type_1.t.generic(User),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "user", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [live_database_1.LiveDatabase, orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [User]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const testController = client.controller('test');
    await testController.start();
    const user = await testController.user();
    globals_1.expect(user.value.name).toBe('peter');
    globals_1.expect(user.value.connections).toBe(0);
    testController.increase(1);
    console.log('wait 1');
    await user.nextStateChange;
    globals_1.expect(user.value.name).toBe('peter');
    globals_1.expect(user.value.connections).toBe(1);
    testController.increase(-5);
    console.log('wait 2');
    await user.nextStateChange;
    globals_1.expect(user.value.name).toBe('peter');
    globals_1.expect(user.value.connections).toBe(-4);
    testController.increase(1);
    console.log('wait 3');
    await user.nextStateChange;
    globals_1.expect(user.value.name).toBe('peter');
    globals_1.expect(user.value.connections).toBe(-3);
});
globals_1.test('test entity sync list', async () => {
    let User = class User extends UserBase {
    };
    User = __decorate([
        type_1.Entity('user1')
    ], User);
    let TestController = class TestController {
        constructor(liveDatabase, database) {
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(User);
        }
        async users() {
            await this.database.query(User).deleteMany();
            const peter = new User('Peter 1');
            await this.database.persist(peter);
            await this.database.persist(new User('Peter 2'));
            await this.database.persist(new User('Guschdl'));
            await this.database.persist(new User('Ingrid'));
            const items = await this.database.query(User).find();
            globals_1.expect(items[0].id).toBe(peter.id);
            globals_1.expect(items.length).toBe(4);
            setTimeout(async () => {
                console.log('Peter 3 added');
                await this.database.persist(new User('Peter 3'));
            }, 500);
            setTimeout(async () => {
                console.log('Peter 1 patched');
                await this.database.query(User).filter({ id: peter.id }).patchOne({ name: 'Peter patched' });
            }, 1000);
            const collection = await this.liveDatabase.query(User).filter({
                name: { $regex: /Peter/ }
            }).find();
            return collection;
        }
        async addUser(name) {
            await this.database.persist(new User(name));
            return false;
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "users", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "addUser", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [live_database_1.LiveDatabase, orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [User]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const testController = client.controller('test');
    const users = await testController.users();
    globals_1.expect(users.count()).toBe(2);
    globals_1.expect(users.all()[0].name).toBe('Peter 1');
    globals_1.expect(users.all()[1].name).toBe('Peter 2');
    console.log('users.nextStateChange 1');
    await users.nextStateChange;
    globals_1.expect(users.count()).toBe(3);
    globals_1.expect(users.all()[0].name).toBe('Peter 1');
    globals_1.expect(users.all()[1].name).toBe('Peter 2');
    globals_1.expect(users.all()[2].name).toBe('Peter 3');
    console.log('users.nextStateChange 2');
    await users.nextStateChange;
    globals_1.expect(users.count()).toBe(3);
    globals_1.expect(users.all()[0].name).toBe('Peter patched');
    const testController2 = testing.createRpcClient().controller('test');
    testController2.addUser('Peter 20');
    console.log('users.nextStateChange 3');
    await users.nextStateChange;
    globals_1.expect(users.count()).toBe(4);
    await users.unsubscribe();
    //unsubscribe is sent async, so we wait a bit.
    await core_1.sleep(0.1);
    await testController2.addUser('Peter 30');
    await core_1.sleep(0.1);
    //still 4, since we unsubscribed from feed
    globals_1.expect(users.count()).toBe(4);
});
globals_1.test('test entity sync list: remove', async () => {
    let User = class User extends UserBase {
    };
    User = __decorate([
        type_1.Entity('user2')
    ], User);
    let TestController = class TestController {
        constructor(liveDatabase, database) {
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(User);
        }
        async users() {
            await this.database.query(User).deleteMany();
            await this.database.persist(new User('Peter 1'));
            await this.database.persist(new User('Peter 2'));
            return await this.liveDatabase.query(User).filter({
                name: { $regex: /Peter/ }
            }).find();
        }
        async removeAll() {
            await this.database.query(User).deleteMany();
        }
        async remove(id) {
            await this.database.query(User).filter({ id }).deleteOne();
        }
        async addUser(name) {
            const user = new User(name);
            await this.database.persist(user);
            return user.id;
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "users", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "removeAll", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "remove", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "addUser", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [live_database_1.LiveDatabase, orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [User]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const testController = client.controller('test');
    const users = await testController.users();
    globals_1.expect(users.count()).toBe(2);
    globals_1.expect(users.all()[0].name).toBe('Peter 1');
    globals_1.expect(users.all()[1].name).toBe('Peter 2');
    let peter3Id;
    testController.addUser('Peter 3').then(v => peter3Id = v);
    console.log('wait peter3Id');
    await users.nextStateChange;
    globals_1.expect(users.count()).toBe(3);
    //this triggers no nextStateChange as the filter doesn't match
    await testController.addUser('Nix da');
    testController.remove(peter3Id);
    console.log('wait remove peter3Id');
    await users.nextStateChange;
    globals_1.expect(users.count()).toBe(2);
    testController.removeAll();
    console.log('wait removeAll');
    await users.nextStateChange;
    globals_1.expect(users.count()).toBe(0);
    await users.unsubscribe();
    console.log('done');
});
// test('test entity sync item', async () => {
//     @Entity('user3')
//     class User extends UserBase {
//     }
//     @rpc.controller('test')
//     class TestController {
//         constructor(
//             private connection: RpcKernelConnection,
//             private liveDatabase: LiveDatabase,
//             private database: Database
//         ) {
//             this.liveDatabase.enableChangeFeed(User);
//         }
//         @rpc.action()
//         async user(): Promise<EntitySubject<User>> {
//             await this.database.query(User).deleteMany();
//             await this.database.persist(new User('Guschdl'));
//             const peter = new User('Peter 1');
//             await this.database.persist(peter);
//             this.connection.setTimeout(async () => {
//                 await this.database.query(User).filter({id: peter.id}).patchOne({name: 'Peter patched'});
//             }, 20);
//             this.connection.setTimeout(async () => {
//                 await this.database.query(User).filter({id: peter.id}).deleteOne();
//             }, 280);
//             return await this.liveDatabase.query(User).filter({
//                 name: {$regex: /Peter/}
//             }).findOne();
//         }
//     }
//     const {client, close, app} = await createServerClientPair('test entity sync item', [TestController], [User]);
//     const test = client.controller<TestController>('test');
//     {
//         const user = await test.user();
//         expect(user).toBeInstanceOf(EntitySubject);
//         expect(user.getValue()).toBeInstanceOf(User);
//         expect(user.getValue().name).toBe('Peter 1');
//         const userId = user.getValue().id;
//         const entityStorage = app.lastConnectionInjector!.get(EntityStorage);
//         await user.nextStateChange;
//         expect(user.getValue().name).toBe('Peter patched');
//         expect(entityStorage.needsToBeSend(User, userId, 10000)).toBe(true);
//         await user.nextStateChange;
//         expect(user.deleted).toBe(true);
//         // there are two ways to stop syncing that entity:
//         // call user.unsubscribe() or when server sent next(undefined), which means it got deleted.
//         expect(entityStorage.needsToBeSend(User, userId, 10000)).toBe(false);
//     }
//     {
//         const user = await test.user();
//         expect(user).toBeInstanceOf(EntitySubject);
//         expect(user.getValue()).toBeInstanceOf(User);
//         expect(user.getValue().name).toBe('Peter 1');
//         const userId = user.getValue().id;
//         const entityStorage = app.lastConnectionInjector!.get(EntityStorage);
//         expect(entityStorage.needsToBeSend(User, userId, 10000)).toBe(true);
//         //this happens async, since we sent a message to the server that
//         //we want to stop syncing.
//         user.unsubscribe();
//         await sleep(0.1);
//         expect(entityStorage.needsToBeSend(User, userId, 10000)).toBe(false);
//     }
//     await close();
// });
globals_1.test('test entity sync item undefined', async () => {
    let User = class User extends UserBase {
    };
    User = __decorate([
        type_1.Entity('user4')
    ], User);
    let TestController = class TestController {
        constructor(liveDatabase, database) {
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(User);
        }
        async user() {
            await this.database.query(User).deleteMany();
            await this.database.persist(new User('Guschdl'));
            const peter = new User('Peter 1');
            await this.database.persist(peter);
            return await this.liveDatabase.query(User).filter({
                name: { $regex: /Marie/ }
            }).findOneOrUndefined();
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "user", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [live_database_1.LiveDatabase,
            orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [User]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const controller = client.controller('test');
    const user = await controller.user();
    globals_1.expect(user).toBeUndefined();
});
globals_1.test('test entity sync count', async () => {
    let User = class User extends UserBase {
    };
    User = __decorate([
        type_1.Entity('user5')
    ], User);
    let TestController = class TestController {
        constructor(connection, liveDatabase, database) {
            this.connection = connection;
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(User);
        }
        async userCount() {
            await this.database.query(User).deleteMany();
            await this.database.persist(new User('Guschdl'));
            const peter1 = new User('Peter 1');
            this.connection.setTimeout(async () => {
                console.log('add peter1');
                await this.database.persist(peter1);
            }, 100);
            this.connection.setTimeout(async () => {
                console.log('add peter2');
                await this.database.persist(new User('Peter 2'));
            }, 150);
            this.connection.setTimeout(async () => {
                console.log('remove peter1');
                await this.database.query(User).filter({ id: peter1.id }).deleteOne();
            }, 200);
            return await this.liveDatabase.query(User).filter({
                name: { $regex: /Peter/ }
            }).count();
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "userCount", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [rpc_1.RpcKernelConnection,
            live_database_1.LiveDatabase,
            orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [User]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const controller = client.controller('test');
    const result = await controller.userCount();
    let i = 0;
    result.subscribe((next) => {
        console.log('next', i, next);
        if (i === 0) {
            //first count is not found
            globals_1.expect(next).toBe(0);
        }
        if (i === 1) {
            //we created peter 1
            globals_1.expect(next).toBe(1);
        }
        if (i === 2) {
            //we created peter 2
            globals_1.expect(next).toBe(2);
        }
        if (i === 3) {
            //we deleted peter 1
            globals_1.expect(next).toBe(1);
        }
        i++;
    });
    const userCount = new rxjs_1.BehaviorSubject(0);
    globals_1.expect(userCount.getValue()).toBe(0);
    result.subscribe((next) => {
        userCount.next(next);
    });
    await core_rxjs_1.nextValue(userCount);
    globals_1.expect(userCount.getValue()).toBe(0);
    await core_rxjs_1.nextValue(userCount);
    globals_1.expect(userCount.getValue()).toBe(1);
    await core_rxjs_1.nextValue(userCount);
    globals_1.expect(userCount.getValue()).toBe(2);
    await core_rxjs_1.nextValue(userCount);
    globals_1.expect(userCount.getValue()).toBe(1);
    globals_1.expect(i).toBe(4);
});
globals_1.test('test entity collection unsubscribe + findOne', async () => {
    let Job = class Job {
        constructor(name) {
            this.name = name;
            this.id = type_1.uuid();
            this.version = 0;
        }
    };
    __decorate([
        type_1.t.primary.uuid,
        __metadata("design:type", String)
    ], Job.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Job.prototype, "version", void 0);
    Job = __decorate([
        type_1.Entity('jobTest'),
        __param(0, type_1.t),
        __metadata("design:paramtypes", [String])
    ], Job);
    globals_1.expect(type_1.getClassSchema(Job).name).toBe('jobTest');
    let TestController = class TestController {
        constructor(liveDatabase, database) {
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(Job);
        }
        async init() {
            await this.database.query(Job).deleteMany();
            await this.database.persist(new Job('Peter 1'));
            await this.database.persist(new Job('Peter 2'));
            await this.database.persist(new Job('Peter 3'));
            await this.database.persist(new Job('Marie 1'));
            await this.database.persist(new Job('Marie 2'));
        }
        async getJobs() {
            return await this.liveDatabase.query(Job).filter({
                name: { $regex: /Peter/ }
            }).find();
        }
        async addJob(name) {
            return await this.database.persist(new Job(name));
        }
        async getJob(id) {
            return await this.liveDatabase.query(Job).filter({
                id: id
            }).findOne();
        }
        async getJobByName(name) {
            return await this.liveDatabase.query(Job).filter({
                name: name
            }).findOne();
        }
        async rmJobByName(name) {
            await this.database.query(Job).filter({
                name: name
            }).deleteOne();
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "init", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "getJobs", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "addJob", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "getJob", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "getJobByName", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "rmJobByName", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [live_database_1.LiveDatabase,
            orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [Job]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const controller = client.controller('test');
    const entityState = client.actionClient.entityState;
    await controller.init();
    try {
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(0);
        const jobs = await controller.getJobs();
        globals_1.expect(jobs.count()).toBe(3);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        const firstId = jobs.all()[0].id;
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(1);
        const firstJob = await controller.getJob(firstId);
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(2);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        globals_1.expect(firstJob.getValue()).toBeInstanceOf(Job);
        globals_1.expect(jobs.get(firstId) === jobs.all()[0]).toBe(true);
        await firstJob.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(1);
        await jobs.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(0);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(0);
    }
    catch (e) {
        assert_1.fail(e);
    }
    try {
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(0);
        const jobs = await controller.getJobs();
        globals_1.expect(jobs.count()).toBe(3);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        const firstId = jobs.all()[0].id;
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(1);
        const firstJob = await controller.getJob(firstId);
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(2);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        globals_1.expect(firstJob.getValue()).toBeInstanceOf(Job);
        globals_1.expect(jobs.get(firstId) === jobs.all()[0]).toBe(true);
        await jobs.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(1);
        await firstJob.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(0);
    }
    catch (e) {
        assert_1.fail(e);
    }
    try {
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(0);
        const jobs = await controller.getJobs();
        globals_1.expect(jobs.count()).toBe(3);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        const firstId = jobs.all()[0].id;
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(1);
        const firstJob1 = await controller.getJob(firstId);
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(2);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        const firstJob2 = await controller.getJob(firstId);
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(3);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(3);
        const firstJob3 = await controller.getJobByName('Marie 1');
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(3);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(4);
        await firstJob1.unsubscribe();
        await firstJob2.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(4);
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(1);
        controller.addJob('Peter 10');
        await jobs.nextStateChange;
        globals_1.expect(jobs.count()).toBe(4);
        controller.rmJobByName('Peter 10');
        await jobs.nextStateChange;
        globals_1.expect(jobs.count()).toBe(3);
        await jobs.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getForkCount(firstId)).toBe(0);
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(1);
        await firstJob3.unsubscribe();
        globals_1.expect(entityState.getStore(Job).getEntitySubjectCount()).toBe(0);
    }
    catch (e) {
        assert_1.fail(e);
    }
});
globals_1.test('test entity collection pagination', async () => {
    let Item = class Item {
        constructor(name, nr = 0, clazz = 'a', owner) {
            this.name = name;
            this.nr = nr;
            this.clazz = clazz;
            this.owner = owner;
            this.id = type_1.uuid();
            this.version = 0;
        }
    };
    __decorate([
        type_1.t.primary.uuid,
        __metadata("design:type", String)
    ], Item.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Item.prototype, "version", void 0);
    Item = __decorate([
        type_1.Entity('paginate/item'),
        __param(0, type_1.t),
        __param(1, type_1.t),
        __param(2, type_1.t),
        __param(3, type_1.t.uuid),
        __metadata("design:paramtypes", [String, Number, String, String])
    ], Item);
    let TestController = class TestController {
        constructor(connection, liveDatabase, database) {
            this.connection = connection;
            this.liveDatabase = liveDatabase;
            this.database = database;
            this.liveDatabase.enableChangeFeed(Item);
        }
        async init() {
            await this.database.query(Item).deleteMany();
            const promises = [];
            const clazzes = ['a', 'b', 'c'];
            const owners = ['3f63154d-4121-4f5c-a297-afc1f8f453fd', '4c349fe0-fa33-4e10-bb17-e25f13e4740e'];
            for (let i = 0; i < 100; i++) {
                promises.push(this.database.persist(new Item('name_' + i, i, clazzes[i % 3], owners[i % 2])));
            }
            await Promise.all(promises);
            console.log('init done!');
            const c = await this.database.query(Item).count();
            globals_1.expect(c).toBe(100);
        }
        async add(clazz, nr) {
            const item = new Item('name_' + nr, nr, clazz, '3f63154d-4121-4f5c-a297-afc1f8f453fd');
            await this.database.persist(item);
        }
        async remove(name) {
            await this.database.query(Item).filter({ name }).deleteOne();
        }
        async findByClass(clazz) {
            return this.liveDatabase.query(Item)
                .filter({ clazz: { $parameter: 'clazz' } })
                .parameter('clazz', clazz)
                .itemsPerPage(10)
                .page(1)
                .sort({ nr: 'asc' })
                .find();
        }
        async findByOwner(owner) {
            return this.liveDatabase.query(Item)
                .filter({ owner: owner })
                .find();
        }
        async findByOwnerPaged(owner) {
            return this.liveDatabase.query(Item)
                .filter({ owner: owner })
                .itemsPerPage(30)
                .page(2)
                .find();
        }
    };
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "init", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String, Number]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "add", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "remove", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "findByClass", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "findByOwner", null);
    __decorate([
        rpc_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Promise)
    ], TestController.prototype, "findByOwnerPaged", null);
    TestController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [rpc_1.RpcKernelConnection,
            live_database_1.LiveDatabase,
            orm_1.Database])
    ], TestController);
    const testing = createTestingApp({ controllers: [TestController] }, [Item]);
    await testing.startServer();
    const client = testing.createRpcClient();
    const controller = client.controller('test');
    await controller.init();
    console.log('test.init done');
    {
        console.log('findByOwner');
        const items = await controller.findByOwner('3f63154d-4121-4f5c-a297-afc1f8f453fd');
        globals_1.expect(items.getPage()).toBe(1);
        console.log('findByOwner done');
        globals_1.expect(items.count()).toBe(50);
        controller.add('a', 1000);
        console.log('item added a');
        await items.nextStateChange;
        console.log('done');
        globals_1.expect(items.count()).toBe(51);
        controller.remove('name_1000');
        console.log('item removed name_1000');
        await items.nextStateChange;
        console.log('done');
        globals_1.expect(items.count()).toBe(50);
    }
    {
        const items = await controller.findByOwnerPaged('3f63154d-4121-4f5c-a297-afc1f8f453fd');
        console.log('items.model', items.model);
        globals_1.expect(items.getPage()).toBe(2);
        globals_1.expect(items.getTotal()).toBe(50);
        globals_1.expect(items.getItemsPerPage()).toBe(30);
        globals_1.expect(items.count()).toBe(20); //we got 50 total, 30 per page. 30,20, we are at second page.
        controller.add('a', 1001);
        await items.nextStateChange;
        globals_1.expect(items.count()).toBe(21);
        controller.remove('name_1001');
        await items.nextStateChange;
        globals_1.expect(items.count()).toBe(20);
        await items.unsubscribe();
    }
    {
        const items = await controller.findByClass('a');
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getItemsPerPage()).toBe(10);
        globals_1.expect(items.getSort()).toEqual({ "nr": "asc" });
        globals_1.expect(items.getParameter('clazz')).toBe('a');
        globals_1.expect(items.all()[0].nr).toBe(0);
        globals_1.expect(items.all()[9].nr).toBe(27);
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getPages()).toBe(4);
        await items.setParameter('clazz', 'b').apply();
        globals_1.expect(items.getTotal()).toBe(33);
        globals_1.expect(items.getItemsPerPage()).toBe(10);
        globals_1.expect(items.getSort()).toEqual({ "nr": "asc" });
        globals_1.expect(items.getParameter('clazz')).toBe('b');
        globals_1.expect(items.all()[0].nr).toBe(1);
        globals_1.expect(items.all()[9].nr).toBe(28);
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.getTotal()).toBe(33);
        globals_1.expect(items.getPages()).toBe(4);
        await items.unsubscribe();
    }
    {
        const items = await controller.findByClass('a');
        globals_1.expect(items.all()[0].nr).toBe(0);
        globals_1.expect(items.all()[9].nr).toBe(27);
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getPages()).toBe(4);
        console.log('swap order');
        await items.orderByField('nr', 'desc').apply();
        globals_1.expect(items.all()[0].nr).toBe(99);
        globals_1.expect(items.all()[9].nr).toBe(72);
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getPages()).toBe(4);
        console.log('swap order back');
        await items.orderByField('nr').apply();
        globals_1.expect(items.all()[0].nr).toBe(0);
        globals_1.expect(items.all()[9].nr).toBe(27);
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getPages()).toBe(4);
        controller.remove('name_27');
        await items.nextStateChange;
        globals_1.expect(items.all()[0].nr).toBe(0);
        globals_1.expect(items.all()[9].nr).toBe(30);
        globals_1.expect(items.getTotal()).toBe(33);
        globals_1.expect(items.getPages()).toBe(4);
        controller.remove('name_0');
        await items.nextStateChange;
        globals_1.expect(items.all()[0].nr).toBe(3);
        globals_1.expect(items.all()[9].nr).toBe(33);
        globals_1.expect(items.getTotal()).toBe(32);
        globals_1.expect(items.getPages()).toBe(4);
        //shouldn't change anything
        controller.add('a', 101);
        await items.nextStateChange;
        globals_1.expect(items.all()[0].nr).toBe(3);
        globals_1.expect(items.all()[9].nr).toBe(33);
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.getTotal()).toBe(33);
        globals_1.expect(items.getPages()).toBe(4);
        //should change a lot
        controller.add('a', -1);
        await items.nextStateChange;
        globals_1.expect(items.all()[0].nr).toBe(-1);
        globals_1.expect(items.all()[9].nr).toBe(30);
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getPages()).toBe(4);
        await items.setPage(2).apply();
        globals_1.expect(items.count()).toBe(10);
        globals_1.expect(items.all()[0].nr).toBe(33);
        globals_1.expect(items.all()[9].nr).toBe(60);
        globals_1.expect(items.getTotal()).toBe(34);
        globals_1.expect(items.getPages()).toBe(4);
        items.unsubscribe();
    }
});
