// @deepkit/framework/dist/cjs/tests/service-container.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
require("reflect-metadata");
const application_service_container_1 = require("../src/application-service-container");
const rpc_1 = require("@deepkit/rpc");
const app_1 = require("@deepkit/app");
globals_1.test('controller', () => {
    class MyService {
        constructor(text = 'hello') {
            this.text = text;
        }
        getHello() {
            return this.text;
        }
    }
    let MyController = class MyController {
        constructor(myService) {
            this.myService = myService;
        }
        foo() {
            return this.myService.getHello();
        }
    };
    MyController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [MyService])
    ], MyController);
    {
        const myModule = new app_1.AppModule({
            providers: [MyService],
            controllers: [MyController],
        });
        const serviceContainer = new application_service_container_1.ApplicationServiceContainer(myModule);
        const rpcScopedContext = serviceContainer.getRootInjectorContext().createChildScope('rpc');
        const controller = rpcScopedContext.getInjector(0).get(MyController);
        globals_1.expect(controller).toBeInstanceOf(MyController);
        globals_1.expect(controller.foo()).toBe('hello');
    }
});
globals_1.test('controller in module and overwrite service', () => {
    class MyService {
        constructor(text = 'hello') {
            this.text = text;
        }
        getHello() {
            return this.text;
        }
    }
    let MyController = class MyController {
        constructor(myService) {
            this.myService = myService;
        }
        foo() {
            return this.myService.getHello();
        }
    };
    MyController = __decorate([
        rpc_1.rpc.controller('test'),
        __metadata("design:paramtypes", [MyService])
    ], MyController);
    const controllerModule = new app_1.AppModule({
        providers: [MyService],
        controllers: [MyController],
        exports: [
            MyService
        ]
    });
    {
        const myModule = new app_1.AppModule({
            imports: [controllerModule],
        });
        const serviceContainer = new application_service_container_1.ApplicationServiceContainer(myModule);
        const rpcScopedContext = serviceContainer.getRootInjectorContext().createChildScope('rpc');
        const controller = rpcScopedContext.get(MyController);
        globals_1.expect(controller).toBeInstanceOf(MyController);
        globals_1.expect(controller.foo()).toBe('hello');
    }
    {
        const myModule = new app_1.AppModule({
            providers: [
                { provide: MyService, useValue: new MyService('different') }
            ],
            imports: [controllerModule],
        });
        const serviceContainer = new application_service_container_1.ApplicationServiceContainer(myModule);
        const rpcScopedContext = serviceContainer.getRootInjectorContext().createChildScope('rpc');
        const controller = rpcScopedContext.get(MyController);
        globals_1.expect(controller).toBeInstanceOf(MyController);
        globals_1.expect(controller.foo()).toBe('different');
    }
});
