// @deepkit/stopwatch/dist/cjs/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameType = exports.FrameCategory = void 0;
var FrameCategory;
(function (FrameCategory) {
    FrameCategory[FrameCategory["none"] = 0] = "none";
    FrameCategory[FrameCategory["http"] = 1] = "http";
    FrameCategory[FrameCategory["rpc"] = 2] = "rpc";
    FrameCategory[FrameCategory["cli"] = 3] = "cli";
    FrameCategory[FrameCategory["job"] = 4] = "job";
    FrameCategory[FrameCategory["function"] = 5] = "function";
    FrameCategory[FrameCategory["lock"] = 6] = "lock";
    FrameCategory[FrameCategory["workflow"] = 7] = "workflow";
    FrameCategory[FrameCategory["event"] = 8] = "event";
    FrameCategory[FrameCategory["database"] = 9] = "database";
    FrameCategory[FrameCategory["email"] = 10] = "email";
    FrameCategory[FrameCategory["template"] = 11] = "template";
})(FrameCategory = exports.FrameCategory || (exports.FrameCategory = {}));
var FrameType;
(function (FrameType) {
    FrameType[FrameType["start"] = 0] = "start";
    FrameType[FrameType["end"] = 1] = "end";
})(FrameType = exports.FrameType || (exports.FrameType = {}));
