// @deepkit/rpc/dist/cjs/tests/collection.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
require("reflect-metadata");
const type_1 = require("@deepkit/type");
const globals_1 = require("@jest/globals");
const client_direct_1 = require("../src/client/client-direct");
const collection_1 = require("../src/collection");
const decorators_1 = require("../src/decorators");
const kernel_1 = require("../src/server/kernel");
const core_1 = require("@deepkit/core");
globals_1.test('collection basic', () => {
    class Item {
        constructor(id, name) {
            this.id = id;
            this.name = name;
            this.version = 0;
        }
    }
    const collection = new collection_1.Collection(Item);
    const item1 = new Item(1, 'Item 1');
    const item2 = new Item(2, 'Item 2');
    const item3 = new Item(3, 'Item 3');
    collection.set([item1, item2]);
    globals_1.expect(collection.count()).toBe(2);
    globals_1.expect(collection.has(item1.id)).toBe(true);
    globals_1.expect(collection.has(item2.id)).toBe(true);
    globals_1.expect(collection.get(item2.id)).toBe(item2);
    globals_1.expect(collection.has(item3.id)).toBe(false);
    collection.add(item3);
    globals_1.expect(collection.has(item3.id)).toBe(true);
    globals_1.expect(collection.index(item3)).toBe(2);
    globals_1.expect(collection.count()).toBe(3);
    globals_1.expect(collection.getPageOf(item1, 2)).toBe(0);
    globals_1.expect(collection.getPageOf(item2, 2)).toBe(0);
    globals_1.expect(collection.getPageOf(item3, 2)).toBe(1);
    //adding same id just replace it
    collection.add(item3);
    globals_1.expect(collection.has(item3.id)).toBe(true);
    globals_1.expect(collection.index(item3)).toBe(2);
    globals_1.expect(collection.count()).toBe(3);
    globals_1.expect(collection.getPageOf(item3, 2)).toBe(1);
    const map = collection.map();
    globals_1.expect(map.get(item3.id)).toBe(item3);
    globals_1.expect(collection.ids()).toEqual([
        item1.id, item2.id, item3.id
    ]);
    globals_1.expect(collection.empty()).toBe(false);
    collection.reset();
    globals_1.expect(collection.count()).toBe(0);
    globals_1.expect(collection.empty()).toBe(true);
});
globals_1.test('collection state', async () => {
    let MyModel = class MyModel {
        constructor() {
            this.id = 0;
        }
    };
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], MyModel.prototype, "id", void 0);
    MyModel = __decorate([
        type_1.entity.name('collection/simple/model')
    ], MyModel);
    class Controller {
        constructor() {
            this.collection = new collection_1.Collection(MyModel);
        }
        fix() {
            this.collection.set([
                type_1.plainToClass(MyModel, { id: 1 }),
                type_1.plainToClass(MyModel, { id: 2 }),
                type_1.plainToClass(MyModel, { id: 3 }),
            ]);
            this.collection.state.total = 150;
            return this.collection;
        }
        changedModel() {
            const collection = new collection_1.Collection(MyModel);
            collection.model.itemsPerPage = 30;
            collection.model.skip = 30;
            collection.model.limit = 5;
            collection.model.sort = { id: 'asc' };
            return collection;
        }
        add(id) {
            this.collection.add({ id });
        }
        remove(id) {
            this.collection.remove(id);
        }
        set(id) {
            const items = id.map(v => { return { id: v }; });
            this.collection.set(items);
        }
    }
    __decorate([
        decorators_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", collection_1.Collection)
    ], Controller.prototype, "fix", null);
    __decorate([
        decorators_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", collection_1.Collection)
    ], Controller.prototype, "changedModel", null);
    __decorate([
        decorators_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "add", null);
    __decorate([
        decorators_1.rpc.action(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "remove", null);
    __decorate([
        decorators_1.rpc.action(),
        __param(0, type_1.t.array(type_1.t.number)),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Array]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "set", null);
    const kernel = new kernel_1.RpcKernel();
    kernel.registerController('myController', Controller);
    const client = new client_direct_1.DirectClient(kernel);
    const controller = client.controller('myController');
    {
        const c = await controller.changedModel();
        globals_1.expect(c.model.itemsPerPage).toBe(30);
        globals_1.expect(c.model.skip).toBe(30);
        globals_1.expect(c.model.limit).toBe(5);
        globals_1.expect(c.model.sort).toEqual({ id: 'asc' });
    }
    {
        const c = await controller.fix();
        globals_1.expect(c.classType).toBe(MyModel);
        globals_1.expect(c.state.total).toBe(150);
        globals_1.expect(c.all().length).toBe(3);
        globals_1.expect(c.all()[0]).toBeInstanceOf(MyModel);
        globals_1.expect(c.all()[0].id).toBe(1);
        globals_1.expect(c.all()[1]).toBeInstanceOf(MyModel);
        globals_1.expect(c.all()[1].id).toBe(2);
        controller.add(4);
        await c.nextStateChange;
        globals_1.expect(c.all().length).toBe(4);
        globals_1.expect(c.all()[3].id).toBe(4);
        controller.remove(1);
        await c.nextStateChange;
        globals_1.expect(c.all().length).toBe(3);
        globals_1.expect(c.all()[0].id).toBe(2);
        globals_1.expect(c.all()[1].id).toBe(3);
        globals_1.expect(c.all()[2].id).toBe(4);
        controller.set([10, 11,]);
        await c.nextStateChange;
        globals_1.expect(c.all().length).toBe(2);
        globals_1.expect(c.all()[0].id).toBe(10);
        globals_1.expect(c.all()[1].id).toBe(11);
        c.unsubscribe();
        controller.set([55]);
        await core_1.sleep(0.1);
        //should not change anything, since we unsubscribed
        globals_1.expect(c.all().length).toBe(2);
        globals_1.expect(c.all()[0].id).toBe(10);
        globals_1.expect(c.all()[1].id).toBe(11);
    }
});
