// @deepkit/http/dist/cjs/tests/router.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
require("reflect-metadata");
const router_1 = require("../src/router");
const decorator_1 = require("../src/decorator");
const type_1 = require("@deepkit/type");
const http_1 = require("../src/http");
const kernel_1 = require("../src/kernel");
const event_1 = require("@deepkit/event");
const injector_1 = require("@deepkit/injector");
const logger_1 = require("@deepkit/logger");
const model_1 = require("../src/model");
const core_1 = require("@deepkit/core");
const stopwatch_1 = require("@deepkit/stopwatch");
function createHttpKernel(controllers, providers = [], listeners = []) {
    const tagProviders = new injector_1.TagRegistry();
    for (const provider of providers.slice(0)) {
        if (provider instanceof injector_1.TagProvider) {
            providers.unshift(provider.provider);
            tagProviders.tags.push(provider);
        }
    }
    const router = router_1.Router.forControllers(controllers, tagProviders);
    const injector = injector_1.InjectorContext.forProviders([
        { provide: router_1.Router, useValue: router },
        ...controllers,
        ...providers,
        ...listeners,
        http_1.HttpListener,
        { provide: logger_1.Logger, useValue: new logger_1.Logger([new logger_1.ConsoleTransport()]) },
        stopwatch_1.Stopwatch
    ]);
    const eventDispatcher = new event_1.EventDispatcher(injector);
    eventDispatcher.registerListener(http_1.HttpListener);
    for (const listener of listeners)
        eventDispatcher.registerListener(listener);
    return new kernel_1.HttpKernel(router, eventDispatcher, injector, new logger_1.Logger([new logger_1.ConsoleTransport()]), new stopwatch_1.Stopwatch());
}
globals_1.test('router', async () => {
    var _a, _b, _c;
    class Controller {
        helloWorld() {
        }
        hello(name) {
        }
        userStatic(id) {
        }
        userStatic2(id, id2) {
        }
        static() {
        }
    }
    __decorate([
        decorator_1.http.GET(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "helloWorld", null);
    __decorate([
        decorator_1.http.GET(':name'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "hello", null);
    __decorate([
        decorator_1.http.GET('/user/:id/static'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "userStatic", null);
    __decorate([
        decorator_1.http.GET('/user2/:id/static/:id2'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String, String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "userStatic2", null);
    __decorate([
        decorator_1.http.GET('/static'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "static", null);
    const router = router_1.Router.forControllers([Controller]);
    globals_1.expect((_a = (await router.resolve('GET', '/'))) === null || _a === void 0 ? void 0 : _a.routeConfig.action).toMatchObject({ controller: Controller, methodName: 'helloWorld' });
    globals_1.expect((_b = (await router.resolve('GET', '/peter'))) === null || _b === void 0 ? void 0 : _b.routeConfig.action).toMatchObject({ controller: Controller, methodName: 'hello' });
    globals_1.expect((_c = (await router.resolve('GET', '/peter'))) === null || _c === void 0 ? void 0 : _c.parameters(undefined)).toEqual(['peter']);
    const userStatic = await router.resolve('GET', '/user/1233/static');
    globals_1.expect(userStatic === null || userStatic === void 0 ? void 0 : userStatic.routeConfig.action).toMatchObject({ controller: Controller, methodName: 'userStatic' });
    globals_1.expect(userStatic === null || userStatic === void 0 ? void 0 : userStatic.parameters(undefined)).toEqual(['1233']);
    const userStatic2 = await router.resolve('GET', '/user2/1233/static/123');
    globals_1.expect(userStatic2 === null || userStatic2 === void 0 ? void 0 : userStatic2.routeConfig.action).toMatchObject({ controller: Controller, methodName: 'userStatic2' });
    globals_1.expect(userStatic2 === null || userStatic2 === void 0 ? void 0 : userStatic2.parameters(undefined)).toEqual(['1233', '123']);
});
globals_1.test('router parameters', async () => {
    var _a;
    class Controller {
        string(name) {
            return new http_1.JSONResponse(name);
        }
        number(id) {
            return new http_1.JSONResponse(id);
        }
        boolean(yes) {
            return new http_1.JSONResponse(yes);
        }
        any(path) {
            return new http_1.JSONResponse(path);
        }
    }
    __decorate([
        decorator_1.http.GET('/user/:name'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "string", null);
    __decorate([
        decorator_1.http.GET('/user-id/:id'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "number", null);
    __decorate([
        decorator_1.http.GET('/boolean/:yes'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Boolean]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "boolean", null);
    __decorate([
        decorator_1.http.GET(':path').regexp('path', '.*'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "any", null);
    const router = router_1.Router.forControllers([Controller]);
    globals_1.expect((_a = (await router.resolve('GET', '/user/peter'))) === null || _a === void 0 ? void 0 : _a.routeConfig.action).toMatchObject({ controller: Controller, methodName: 'string' });
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/user/peter')).toBe('peter');
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/user-id/123')).toBe(123);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/user-id/asd')).toMatchObject({ message: 'Validation failed: id(invalid_number): No valid number given, got NaN' });
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/boolean/1')).toBe(true);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/boolean/false')).toBe(false);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/any')).toBe('any');
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/any/path')).toBe('any/path');
});
globals_1.test('router HttpRequest', async () => {
    class Controller {
        anyReq(req, path) {
            return [req.url, path];
        }
    }
    __decorate([
        decorator_1.http.GET(':path').regexp('path', '.*'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [model_1.HttpRequest, String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "anyReq", null);
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/req/any/path')).toEqual(['/req/any/path', 'req/any/path']);
});
globals_1.test('router parameterResolver', async () => {
    class User {
        constructor(username) {
            this.username = username;
        }
    }
    class Group {
        constructor(name) {
            this.name = name;
        }
    }
    class Controller {
        route1(user) {
            return [user.username];
        }
        route2(user) {
            return [user.username];
        }
        route3(user, group) {
            return [user.username, group.name];
        }
    }
    __decorate([
        decorator_1.http.GET('user/:username'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [User]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "route1", null);
    __decorate([
        decorator_1.http.GET('invalid'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [User]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "route2", null);
    __decorate([
        decorator_1.http.GET('user/:user/group/:group'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [User, Group]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "route3", null);
    class UserResolver {
        resolve(context) {
            const value = context.value || context.parameters.username;
            if (!value)
                throw new Error('No value specified');
            return new User(value);
        }
    }
    class GroupResolver {
        resolve(context) {
            if (!context.value)
                throw new Error('No value specified');
            return new Group(context.value);
        }
    }
    const httpKernel = createHttpKernel([Controller], [router_1.RouteParameterResolverTag.provide(UserResolver).forClassType(User), router_1.RouteParameterResolverTag.provide(GroupResolver).forClassType(Group)]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/user/peter')).toEqual(['peter']);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/user/peter/group/a')).toEqual(['peter', 'a']);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/invalid')).toEqual('Internal error');
});
globals_1.test('router body', async () => {
    class Body {
    }
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], Body.prototype, "username", void 0);
    class Controller {
        anyReq(body, req) {
            return [body.username, body instanceof Body, req.url];
        }
    }
    __decorate([
        decorator_1.http.POST(),
        __param(0, decorator_1.http.body()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Body, model_1.HttpRequest]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "anyReq", null);
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('POST', '/', { username: 'Peter' })).toEqual(['Peter', true, '/']);
});
globals_1.test('router body double', async () => {
    class Body {
    }
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], Body.prototype, "username", void 0);
    class Controller {
        anyReq(body, body2, req) {
            return [body2.username, body2 instanceof Body && body instanceof Body, req.url];
        }
    }
    __decorate([
        decorator_1.http.POST(),
        __param(0, decorator_1.http.body()),
        __param(1, decorator_1.http.body()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Body, Body, model_1.HttpRequest]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "anyReq", null);
    const httpData = decorator_1.httpClass._fetch(Controller);
    if (!httpData)
        throw new Error('httpClass undefined');
    const action = [...httpData.getActions()][0];
    globals_1.expect(action.methodName).toBe('anyReq');
    globals_1.expect(action.httpMethod).toBe('POST');
    globals_1.expect(action.parameters['body']).not.toBeUndefined();
    globals_1.expect(action.parameters['body'].name).toBe('body');
    globals_1.expect(action.parameters['body2']).not.toBeUndefined();
    globals_1.expect(action.parameters['body2'].name).toBe('body2');
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('POST', '/', { username: 'Peter' })).toEqual(['Peter', true, '/']);
});
globals_1.test('router groups', async () => {
    {
        class Controller {
            a() {
            }
            b() {
            }
            c() {
            }
        }
        __decorate([
            decorator_1.http.GET('a').group('a'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], Controller.prototype, "a", null);
        __decorate([
            decorator_1.http.GET('b'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], Controller.prototype, "b", null);
        __decorate([
            decorator_1.http.GET('c').group('c'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], Controller.prototype, "c", null);
        const httpData = decorator_1.httpClass._fetch(Controller);
        if (!httpData)
            throw new Error('httpClass undefined');
        globals_1.expect(httpData.getAction('a').groups).toEqual(['a']);
        globals_1.expect(httpData.getAction('b').groups).toEqual([]);
        globals_1.expect(httpData.getAction('c').groups).toEqual(['c']);
    }
    {
        let Controller = class Controller {
            a() {
            }
            b() {
            }
            c() {
            }
        };
        __decorate([
            decorator_1.http.GET('a').group('a'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], Controller.prototype, "a", null);
        __decorate([
            decorator_1.http.GET('b'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], Controller.prototype, "b", null);
        __decorate([
            decorator_1.http.GET('c').group('c'),
            __metadata("design:type", Function),
            __metadata("design:paramtypes", []),
            __metadata("design:returntype", void 0)
        ], Controller.prototype, "c", null);
        Controller = __decorate([
            decorator_1.http.groupAll('all')
        ], Controller);
        const httpData = decorator_1.httpClass._fetch(Controller);
        if (!httpData)
            throw new Error('httpClass undefined');
        globals_1.expect(httpData.getAction('a').groups).toEqual(['a', 'all']);
        globals_1.expect(httpData.getAction('b').groups).toEqual(['all']);
        globals_1.expect(httpData.getAction('c').groups).toEqual(['c', 'all']);
    }
    globals_1.expect(() => {
        let ControllerC = class ControllerC {
        };
        ControllerC = __decorate([
            decorator_1.http.group('all')
        ], ControllerC);
    }).toThrow('Property decorators can only be used on class properties');
});
globals_1.test('router query', async () => {
    class Controller {
        anyReq(test) {
            return test;
        }
    }
    __decorate([
        decorator_1.http.GET('my-action'),
        __param(0, decorator_1.http.query().optional),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Number]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "anyReq", null);
    const httpData = decorator_1.httpClass._fetch(Controller);
    if (!httpData)
        throw new Error('httpClass undefined');
    const action = [...httpData.getActions()][0];
    globals_1.expect(action.methodName).toBe('anyReq');
    globals_1.expect(action.httpMethod).toBe('GET');
    globals_1.expect(action.parameters['test']).not.toBeUndefined();
    globals_1.expect(action.parameters['test'].name).toBe('test');
    globals_1.expect(action.parameters['test'].type).toBe('query');
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/my-action?test=123')).toEqual(123);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/my-action')).toEqual(undefined);
});
globals_1.test('router query all', async () => {
    class AnyReqQuery {
    }
    __decorate([
        type_1.t.optional,
        __metadata("design:type", String)
    ], AnyReqQuery.prototype, "test", void 0);
    __decorate([
        type_1.t.optional,
        __metadata("design:type", String)
    ], AnyReqQuery.prototype, "filter", void 0);
    __decorate([
        type_1.t.optional,
        __metadata("design:type", Number)
    ], AnyReqQuery.prototype, "page", void 0);
    class Controller {
        anyReq(anyReqQuery) {
            return anyReqQuery;
        }
    }
    __decorate([
        decorator_1.http.GET('my-action'),
        __param(0, decorator_1.http.queries()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [AnyReqQuery]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "anyReq", null);
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/my-action?test=123')).toEqual({ test: '123' });
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/my-action')).toEqual({});
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/my-action?filter=page&page=5')).toEqual({ filter: 'page', page: 5 });
});
globals_1.test('serializer options', async () => {
    class User {
    }
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], User.prototype, "username", void 0);
    __decorate([
        type_1.t.group('sensitive'),
        __metadata("design:type", String)
    ], User.prototype, "password", void 0);
    class Controller {
        anyReq() {
            return { username: 'Peter', password: 'secret' };
        }
    }
    __decorate([
        decorator_1.http.GET().serialization({ groupsExclude: ['sensitive'] }),
        type_1.t.type(User),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "anyReq", null);
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/')).toEqual({ username: 'Peter' });
});
globals_1.test('hook after serializer', async () => {
    class User {
    }
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], User.prototype, "username", void 0);
    __decorate([
        type_1.t.group('sensitive'),
        __metadata("design:type", String)
    ], User.prototype, "password", void 0);
    class Controller {
        async anyReq() {
            await core_1.sleep(0.1);
            return { username: 'Peter', password: 'secret' };
        }
    }
    __decorate([
        decorator_1.http.GET().serialization({ groupsExclude: ['sensitive'] }),
        type_1.t.type(User),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Controller.prototype, "anyReq", null);
    class Listener {
        onResponse(event) {
            event.result = { processingTime: event.controllerActionTime, data: event.result };
        }
    }
    __decorate([
        event_1.eventDispatcher.listen(http_1.httpWorkflow.onResponse),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Listener.prototype, "onResponse", null);
    const httpKernel = createHttpKernel([Controller], [], [Listener]);
    const result = await httpKernel.handleRequestFor('GET', '/');
    globals_1.expect(result.data).toEqual({ username: 'Peter' });
    globals_1.expect(result.processingTime).toBeGreaterThanOrEqual(99);
});
globals_1.test('custom request handling', async () => {
    class Listener {
        onRouteNotFound(event) {
            //CORS requirement for example
            if (event.request.method === 'OPTIONS')
                event.send(new http_1.JSONResponse(true, 200));
        }
    }
    __decorate([
        event_1.eventDispatcher.listen(http_1.httpWorkflow.onRouteNotFound),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Listener.prototype, "onRouteNotFound", null);
    const httpKernel = createHttpKernel([], [], [Listener]);
    globals_1.expect(await httpKernel.handleRequestFor('OPTIONS', '/')).toBe(true);
});
globals_1.test('promise serializer', async () => {
    class Controller {
        async anyReq1() {
            return "test";
        }
        async anyReq2() {
            return 1;
        }
    }
    __decorate([
        decorator_1.http.GET('1'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Controller.prototype, "anyReq1", null);
    __decorate([
        decorator_1.http.GET('2'),
        type_1.t.string,
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Promise)
    ], Controller.prototype, "anyReq2", null);
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/1')).toBe("test");
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/2')).toBe("1");
});
globals_1.test('unions', async () => {
    class Controller {
        list(page) {
            return page;
        }
    }
    __decorate([
        decorator_1.http.GET('/list'),
        __param(0, decorator_1.http.query()),
        __param(0, type_1.t.union(type_1.t.number, type_1.t.boolean)),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "list", null);
    const httpKernel = createHttpKernel([Controller]);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/list?page=1')).toEqual(1);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/list?page=2222')).toEqual(2222);
    globals_1.expect(await httpKernel.handleRequestFor('GET', '/list?page=false')).toEqual(false);
});
globals_1.test('router dotToUrlPath', () => {
    globals_1.expect(router_1.dotToUrlPath('peter')).toBe('peter');
    globals_1.expect(router_1.dotToUrlPath('foo.bar')).toBe('foo[bar]');
    globals_1.expect(router_1.dotToUrlPath('foo.bar.deep')).toBe('foo[bar][deep]');
    globals_1.expect(router_1.dotToUrlPath('foo.bar.deep.very')).toBe('foo[bar][deep][very]');
});
globals_1.test('router url resolve', async () => {
    class AnyReqQuery {
    }
    __decorate([
        type_1.t.optional,
        __metadata("design:type", String)
    ], AnyReqQuery.prototype, "test", void 0);
    __decorate([
        type_1.t.optional,
        __metadata("design:type", String)
    ], AnyReqQuery.prototype, "filter", void 0);
    __decorate([
        type_1.t.optional,
        __metadata("design:type", Number)
    ], AnyReqQuery.prototype, "page", void 0);
    class Controller {
        first() {
            return '';
        }
        second(peter) {
            return '';
        }
        secondQuery(peter) {
            return '';
        }
        secondQuery2(peter) {
            return '';
        }
        third(params) {
            return params;
        }
        third2(params) {
            return params;
        }
    }
    __decorate([
        decorator_1.http.GET('').name('first'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "first", null);
    __decorate([
        decorator_1.http.GET(':peter').name('second'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "second", null);
    __decorate([
        decorator_1.http.GET('').name('secondQuery'),
        __param(0, decorator_1.http.query()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "secondQuery", null);
    __decorate([
        decorator_1.http.GET('').name('secondQuery2'),
        __param(0, decorator_1.http.query('changed')),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "secondQuery2", null);
    __decorate([
        decorator_1.http.GET('third').name('third'),
        __param(0, decorator_1.http.queries()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [AnyReqQuery]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "third", null);
    __decorate([
        decorator_1.http.GET('third2').name('third2'),
        __param(0, decorator_1.http.queries('deep')),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [AnyReqQuery]),
        __metadata("design:returntype", void 0)
    ], Controller.prototype, "third2", null);
    const router = router_1.Router.forControllers([Controller]);
    globals_1.expect(router.resolveUrl('first')).toBe('/');
    globals_1.expect(router.resolveUrl('second', { peter: 'foo' })).toBe('/foo');
    globals_1.expect(router.resolveUrl('secondQuery', { peter: 'foo' })).toBe('/?peter=foo');
    globals_1.expect(router.resolveUrl('secondQuery2', { peter: 'foo' })).toBe('/?changed=foo');
    globals_1.expect(router.resolveUrl('third', {})).toBe('/third');
    globals_1.expect(router.resolveUrl('third', { params: { test: 123 } })).toBe('/third?test=123');
    globals_1.expect(router.resolveUrl('third', { params: { test: 123, filter: 'peter' } })).toBe('/third?test=123&filter=peter');
    globals_1.expect(router.resolveUrl('third2', {})).toBe('/third2');
    globals_1.expect(router.resolveUrl('third2', { params: { test: 123 } })).toBe('/third2?deep[test]=123');
    globals_1.expect(router.resolveUrl('third2', { params: { test: 123, filter: 'peter' } })).toBe('/third2?deep[test]=123&deep[filter]=peter');
});
