// @deepkit/type/dist/esm/src/jit-validation-registry.js
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
import { reserveVariable } from './serializer-compiler';
export const validationRegistry = new Map();
export function registerCheckerCompiler(type, compiler) {
    validationRegistry.set(type, compiler);
}
export function executeCheckerCompiler(path, rootContext, jitStack, compiler, getter, property) {
    const res = compiler(getter, property, {
        reserveVariable: (name) => {
            return reserveVariable(rootContext, name);
        },
        context: rootContext,
        path: path,
        raise: (code, message) => {
            return `_errors.push(new ValidationError(${path}, ${JSON.stringify(code)}, ${JSON.stringify(message)}))`;
        },
    }, jitStack);
    if ('string' === typeof res) {
        return res;
    }
    else {
        if (res.context instanceof Map) {
            for (const [k, v] of res.context.entries()) {
                rootContext.set(k, v);
            }
        }
        else {
            for (const i in res.context) {
                if (!res.context.hasOwnProperty(i))
                    continue;
                rootContext.set(i, res.context[i]);
            }
        }
        return res.template;
    }
}
