// @deepkit/mongo/dist/cjs/src/client/host.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Host = void 0;
class Host {
    constructor(hostname, port = 27017) {
        this.hostname = hostname;
        this.port = port;
        this.type = 0 /* unknown */;
        this.connections = [];
    }
    get id() {
        return `${this.hostname}:${this.port}`;
    }
    isWritable() {
        return this.type === 2 /* primary */ || this.type === 1 /* standalone */ || this.type === 4 /* mongos */;
    }
    isSecondary() {
        return this.type === 3 /* secondary */;
    }
    isReadable() {
        return this.type === 2 /* primary */ || this.type === 1 /* standalone */
            || this.type === 4 /* mongos */ || this.type === 3 /* secondary */;
    }
    setType(type) {
        if (this.type !== type) {
            //type changed. Should we do anything special?
        }
        this.type = type;
        this.typeSetAt = new Date;
    }
    getType() {
        return this.type;
    }
    getTypeFromIsMasterResult(isMasterCmdResult) {
        if (!isMasterCmdResult || !isMasterCmdResult.ok)
            return 0 /* unknown */;
        if (isMasterCmdResult.isreplicaset)
            return 7 /* ghost */;
        if (isMasterCmdResult.ismaster && isMasterCmdResult.msg === 'isdbgrid')
            return 4 /* mongos */;
        if (isMasterCmdResult.setName) {
            if (isMasterCmdResult.hidden)
                return 6 /* other */;
            if (isMasterCmdResult.ismaster)
                return 2 /* primary */;
            if (isMasterCmdResult.secondary)
                return 3 /* secondary */;
            if (isMasterCmdResult.arbiterOnly)
                return 5 /* arbiter */;
            return 6 /* other */;
        }
        return 1 /* standalone */;
    }
}
exports.Host = Host;
