// @deepkit/template/dist/esm/tests/simple.js
import { createElement as _createElement } from "..";
import { jsx as _jsx, jsxs as _jsxs } from "../jsx-runtime";
import { html } from '../src/template';
export function simpleOnlyAttributes() {
    return _jsx("div", { id: "123" }, void 0);
}
export function simpleOnlyChildren() {
    return _jsx("div", { children: "Test" }, void 0);
}
export function simple1() {
    return _jsx("div", Object.assign({ id: "123" }, { children: "Test" }), void 0);
}
export function simple2() {
    return _jsx("div", Object.assign({ id: "123" }, { children: _jsx("b", { children: "strong" }, void 0) }), void 0);
}
export function simple3() {
    return _jsxs("div", Object.assign({ id: "123" }, { children: [_jsx("b", { children: "strong" }, void 0), _jsx("b", { children: "strong2" }, void 0)] }), void 0);
}
export function simple4() {
    return _jsx("div", Object.assign({ id: "123", class: "active" }, { children: _jsxs("div", { children: [_jsx("b", { children: "strong" }, void 0), _jsx("b", { children: "strong2" }, void 0)] }, void 0) }), void 0);
}
export function spread() {
    const props = { foo: true, bar: false };
    return _createElement("div", { ...props, key: "myKey" },
        _jsx("b", { children: "TestBold" }, void 0),
        "Test");
}
function Website(props) {
    return _jsxs("html", { children: [_jsx("header", { children: _jsx("meta", Object.assign({ name: "title" }, { children: "My Website" }), void 0) }, void 0), _jsx("body", { children: _jsx("div", { children: props.children }, void 0) }, void 0)] }, void 0);
}
export function simple5(name) {
    return _jsxs(Website, { children: [_jsx("h1", { children: "Hi" }, void 0), _jsxs("p", { children: ["How are you, ", name, "?"] }, void 0)] }, void 0);
}
export function simpleHtmlInjected() {
    const myHtml = '<strong>MyHTML</strong>';
    return _jsx("div", { children: myHtml }, void 0);
}
export function simpleHtmlInjectedValid() {
    const myHtml = '<strong>MyHTML</strong>';
    return _jsx("div", { children: html(myHtml) }, void 0);
}
