// @deepkit/framework/dist/esm/src/broker/broker.module.js
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { AppLocker } from './app-locker';
import { AppModule } from '@deepkit/app';
import { inject, injectable } from '@deepkit/injector';
import { eventDispatcher } from '@deepkit/event';
import { onServerMainBootstrap, onServerMainShutdown } from '../application-server';
import { brokerConfig } from './broker.config';
import { Broker, BrokerServer } from './broker';
let BrokerListener = class BrokerListener {
    constructor(broker, brokerServer, startOnBootstrap) {
        this.broker = broker;
        this.brokerServer = brokerServer;
        this.startOnBootstrap = startOnBootstrap;
    }
    async onMainBootstrap() {
        if (this.startOnBootstrap) {
            await this.brokerServer.start();
        }
    }
    async onMainShutdown() {
        if (this.startOnBootstrap) {
            this.brokerServer.close();
        }
        await this.broker.disconnect();
    }
};
__decorate([
    eventDispatcher.listen(onServerMainBootstrap),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], BrokerListener.prototype, "onMainBootstrap", null);
__decorate([
    eventDispatcher.listen(onServerMainShutdown),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], BrokerListener.prototype, "onMainShutdown", null);
BrokerListener = __decorate([
    injectable(),
    __param(2, inject(brokerConfig.token('startOnBootstrap'))),
    __metadata("design:paramtypes", [Broker,
        BrokerServer, Boolean])
], BrokerListener);
export { BrokerListener };
export const BrokerModule = new AppModule({
    listeners: [
        BrokerListener
    ],
    config: brokerConfig,
    providers: [
        Broker,
        AppLocker,
        BrokerServer,
    ],
    exports: [
        Broker,
        AppLocker,
        BrokerServer,
    ]
}, 'broker').forRoot();
