// @deepkit/core-rxjs/dist/cjs/src/utils.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.tearDown = exports.promiseToObservable = exports.observableToPromise = exports.nextValue = exports.subscriptionToPromise = exports.Subscriptions = exports.isBehaviorSubject = exports.isSubject = exports.AsyncSubscription = void 0;
const rxjs_1 = require("rxjs");
const core_1 = require("@deepkit/core");
const operators_1 = require("rxjs/operators");
class AsyncSubscription {
    constructor(cb) {
        this.cb = cb;
        this.unsubscribed = false;
    }
    async unsubscribe() {
        if (this.unsubscribed)
            return;
        this.unsubscribed = true;
        await this.cb();
    }
}
exports.AsyncSubscription = AsyncSubscription;
function isSubject(v) {
    return v && core_1.isFunction(v.unsubscribe) && rxjs_1.isObservable(v);
}
exports.isSubject = isSubject;
function isBehaviorSubject(v) {
    return v && core_1.isFunction(v.getValue) && isSubject(v);
}
exports.isBehaviorSubject = isBehaviorSubject;
/**
 * RXJS subscription collection, to easily collect multiple subscriptions and unsubscribe all at once.
 * Added subscriptions are automatically removed when they get unsubscribed.
 *
 * @example
 * ```typescript
 * const subs = new Subscriptions();
 *
 * subs.add = new Subscription(() => {});
 * subs.add = observeable.subscribe((next) => {});
 *
 * subs.unsubscribe();
 * ```
 */
class Subscriptions {
    constructor(teardown) {
        this.teardown = teardown;
        this.list = [];
    }
    set add(v) {
        this.list.push(v);
        v.add(() => {
            core_1.arrayRemoveItem(this.list, v);
        });
    }
    unsubscribe() {
        //it's important to work on a array copy, since unsubscribe() modifies directly this.list
        for (const sub of this.list.slice(0)) {
            sub.unsubscribe();
        }
        if (this.teardown) {
            this.teardown();
        }
        this.list.splice(0, this.list.length);
    }
}
exports.Subscriptions = Subscriptions;
function subscriptionToPromise(subscription) {
    return new Promise((resolve) => {
        const sub = subscription.add(() => {
            resolve();
            sub.unsubscribe();
        });
    });
}
exports.subscriptionToPromise = subscriptionToPromise;
function nextValue(o) {
    if (core_1.isFunction(o.getValue)) { //BehaviorSubject
        return o.pipe(operators_1.skip(1)).pipe(operators_1.first()).toPromise();
    }
    return o.pipe(operators_1.first()).toPromise();
}
exports.nextValue = nextValue;
function observableToPromise(o, next) {
    const stack = core_1.createStack();
    return new Promise((resolve, reject) => {
        let last;
        o.subscribe((data) => {
            if (next) {
                next(data);
            }
            last = data;
        }, (error) => {
            core_1.mergeStack(error, stack);
            reject(error);
        }, () => {
            resolve(last);
        });
    });
}
exports.observableToPromise = observableToPromise;
function promiseToObservable(o) {
    const stack = core_1.createStack();
    return new rxjs_1.Observable((observer) => {
        try {
            core_1.mergePromiseStack(o(), stack).then((data) => {
                observer.next(data);
                observer.complete();
            }, (error) => {
                observer.error(error);
            });
        }
        catch (error) {
            observer.error(error);
        }
    });
}
exports.promiseToObservable = promiseToObservable;
async function tearDown(teardown) {
    if ('function' === typeof teardown) {
        await teardown();
    }
    else if ('object' === typeof teardown && teardown.unsubscribe) {
        await teardown.unsubscribe();
    }
}
exports.tearDown = tearDown;
