// @deepkit/framework/dist/esm/src/application-service-container.js
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
import { isProvided, ServiceContainer } from '@deepkit/app';
import { InjectorContext } from '@deepkit/injector';
import { rpcClass } from '@deepkit/rpc';
import { httpClass, HttpControllers } from '@deepkit/http';
export class RpcControllers {
    constructor() {
        this.controllers = new Map();
    }
    resolveController(name) {
        const classType = this.controllers.get(name);
        if (!classType)
            throw new Error(`Controller not found for ${name}`);
        return classType;
    }
}
export class ApplicationServiceContainer extends ServiceContainer {
    constructor() {
        super(...arguments);
        this.rpcControllers = new RpcControllers;
        this.httpControllers = new HttpControllers([]);
    }
    process() {
        if (this.rootContext)
            return;
        this.providers.push({ provide: HttpControllers, useValue: this.httpControllers });
        this.providers.push({ provide: RpcControllers, useValue: this.rpcControllers });
        this.providers.push({ provide: ApplicationServiceContainer, useValue: this });
        return super.process();
    }
    setupController(providers, controller, context) {
        const rpcConfig = rpcClass._fetch(controller);
        if (rpcConfig) {
            if (!isProvided(providers, controller))
                providers.unshift({ provide: controller, scope: 'rpc' });
            controller[InjectorContext.contextSymbol] = context;
            this.rpcControllers.controllers.set(rpcConfig.getPath(), controller);
        }
        const httpConfig = httpClass._fetch(controller);
        if (httpConfig) {
            if (!isProvided(providers, controller))
                providers.unshift({ provide: controller, scope: 'http' });
            controller[InjectorContext.contextSymbol] = context;
            this.httpControllers.add(controller);
        }
        super.setupController(providers, controller, context);
    }
}
