// @deepkit/framework/dist/esm/src/testing.js
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
import { BrokerKernel } from '@deepkit/broker';
import { ConsoleTransport, Logger, MemoryLoggerTransport } from '@deepkit/logger';
import { Database, DatabaseRegistry, MemoryDatabaseAdapter } from '@deepkit/orm';
import { Application } from './application';
import { ApplicationServer } from './application-server';
import { Broker, BrokerServer, DirectBroker } from './broker/broker';
import { injectorReference } from '@deepkit/injector';
import { AppModule } from '@deepkit/app';
import { WebMemoryWorkerFactory, WebWorkerFactory } from './worker';
import { HttpKernel, HttpResponse } from '@deepkit/http';
export class TestHttpResponse extends HttpResponse {
    constructor() {
        super(...arguments);
        this.body = Buffer.alloc(0);
    }
    write(chunk, encoding, callback) {
        if (typeof encoding === 'function') {
            callback = encoding;
            encoding = null;
        }
        if (chunk) {
            if ('string' === typeof chunk) {
                chunk = Buffer.from(chunk, encoding || 'utf8');
            }
            this.body = Buffer.concat([this.body, chunk]);
        }
        if (callback)
            callback();
        return true;
    }
    end(chunk, encoding, callback) {
        if (typeof chunk === 'function') {
            callback = chunk;
            chunk = null;
            encoding = null;
        }
        else if (typeof encoding === 'function') {
            callback = encoding;
            encoding = null;
        }
        if (chunk) {
            if ('string' === typeof chunk) {
                chunk = Buffer.from(chunk, encoding || 'utf8');
            }
            this.body = Buffer.concat([this.body, chunk]);
        }
        if (callback)
            callback();
    }
}
export class TestingFacade {
    constructor(app) {
        this.app = app;
    }
    async startServer() {
        await this.app.get(ApplicationServer).start();
    }
    async stopServer() {
        await this.app.get(ApplicationServer).close();
    }
    async request(requestBuilder) {
        const request = requestBuilder.build();
        const response = new TestHttpResponse(request);
        await this.app.get(HttpKernel).handleRequest(request, response);
        return response;
    }
    createRpcClient() {
        return this.app.get(ApplicationServer).createClient();
    }
}
export class BrokerMemoryServer extends BrokerServer {
    constructor() {
        super(...arguments);
        this.kernel = new BrokerKernel();
    }
    async start() {
    }
    async stop() {
    }
}
/**
 * Creates a new Application instance, but with kernel services in place that work in memory.
 * For example RPC/Broker/HTTP communication without TCP stack. Logger uses MemoryLogger.
 */
export function createTestingApp(optionsOrModule, entities) {
    const module = optionsOrModule instanceof AppModule ? optionsOrModule : new AppModule(optionsOrModule);
    module.setupProvider(Logger).removeTransport(injectorReference(ConsoleTransport));
    module.setupProvider(Logger).addTransport(injectorReference(MemoryLoggerTransport));
    const providers = [
        { provide: WebWorkerFactory, useClass: WebMemoryWorkerFactory },
        { provide: BrokerServer, useClass: BrokerMemoryServer },
        MemoryLoggerTransport,
        {
            provide: Broker, deps: [BrokerServer], useFactory: (server) => {
                return new DirectBroker(server.kernel);
            }
        },
    ];
    if (entities) {
        providers.push({ provide: Database, useValue: new Database(new MemoryDatabaseAdapter, entities) });
        module.setupProvider(DatabaseRegistry).addDatabase(Database);
    }
    return new TestingFacade(new Application(module, providers));
}
