// @deepkit/mongo/dist/esm/benchmarks/bench-connection.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { MongoClient as OriMongoClient } from 'mongodb';
import { Entity, f } from '@deepkit/type';
import { MongoClient } from '../src/client/client';
import { DeleteCommand } from '../src/client/command/delete';
import { InsertCommand } from '../src/client/command/insert';
import { FindCommand } from '../src/client/command/find';
import { bench } from './utils';
let User = class User {
    constructor(id, name) {
        this.id = id;
        this.name = name;
        this.tags = [];
        this.priority = 0;
    }
};
__decorate([
    f.mongoId.primary,
    __metadata("design:type", String)
], User.prototype, "_id", void 0);
__decorate([
    f,
    __metadata("design:type", Boolean)
], User.prototype, "ready", void 0);
__decorate([
    f.array(f.string),
    __metadata("design:type", Array)
], User.prototype, "tags", void 0);
__decorate([
    f,
    __metadata("design:type", Number)
], User.prototype, "priority", void 0);
User = __decorate([
    Entity('user'),
    __param(0, f),
    __param(1, f),
    __metadata("design:paramtypes", [Number, String])
], User);
export { User };
async function main() {
    await bench(10000, 'reference', async () => {
    });
    const items = [];
    const count = 10000;
    for (let i = 1; i <= count; i++) {
        const user = new User(i, 'Peter ' + i);
        user.ready = true;
        user.priority = 5;
        user.tags = ['a', 'b', 'c'];
        items.push(user);
    }
    const oriMongoClient = await OriMongoClient.connect(`mongodb://127.0.0.1/benchmark-a`, {
        useUnifiedTopology: true,
        useNewUrlParser: true,
    });
    const collection = oriMongoClient.db('benchmark-a').collection('user');
    await collection.deleteMany({});
    const client = new MongoClient('mongodb://127.0.0.1/benchmark-a');
    await collection.deleteMany({});
    await bench(1, 'native mongodb insert 10k', async () => {
        await collection.insertMany(items);
    });
    await collection.deleteMany({});
    for (const item of items) {
        item._id = undefined;
    }
    await client.execute(new DeleteCommand(User, {}));
    await bench(1, 'connection2 insert 10k', async () => {
        await client.execute(new InsertCommand(User, items));
    });
    const dbItems = await client.execute(new FindCommand(User, {}));
    console.log('dbItems', dbItems.length);
    if (dbItems.length !== count)
        throw new Error(`Wrong db items count. got ${dbItems.length}, but expected ${count}`);
    await bench(10, 'native mongodb 10k items', async () => {
        const items = await collection.find({}).batchSize(1000).toArray();
    });
    await bench(10, 'connection2 find 10k items', async () => {
        await client.execute(new FindCommand(User, {}));
    });
    await bench(3000, 'native mongodb 1 item', async () => {
        const item = await collection.find({}).limit(1).toArray();
    });
    await bench(10000, 'connection2 find 1 item', async () => {
        await client.execute(new FindCommand(User, {}, undefined, undefined, 1));
    });
    client.close();
    await oriMongoClient.close();
}
main();
