// @deepkit/type/dist/cjs/src/snapshot.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSimplePrimaryKeyHashGenerator = exports.getForeignKeyHash = exports.getPrimaryKeyHashGenerator = exports.getPrimaryKeyExtractor = exports.createSnapshot = exports.getConverterForSnapshot = void 0;
const core_1 = require("@deepkit/core");
const jit_1 = require("./jit");
const json_serializer_1 = require("./json-serializer");
const mapper_1 = require("./mapper");
const model_1 = require("./model");
const serializer_compiler_1 = require("./serializer-compiler");
function createJITConverterForSnapshot(schema, properties, serializerCompilers) {
    const context = new Map();
    const jitStack = new jit_1.JitStack();
    const setProperties = [];
    for (const property of properties) {
        if (property.isParentReference)
            continue;
        if (mapper_1.isExcluded(schema, property.name, 'json'))
            continue;
        if (property.isReference) {
            const referenceCode = [];
            for (const pk of property.getResolvedClassSchema().getPrimaryFields()) {
                referenceCode.push(`
                //createJITConverterForSnapshot ${property.name}->${pk.name} class:snapshot:${property.type} reference
                ${serializer_compiler_1.getDataConverterJS(`_result.${property.name}.${pk.name}`, `_value.${property.name}.${pk.name}`, pk, serializerCompilers, context, jitStack)}
                `);
            }
            setProperties.push(`
            //createJITConverterForSnapshot ${property.name} class:snapshot:${property.type} reference
            if (undefined === _value.${property.name}) {
                _result.${property.name} = null;
            } else if (null === _value.${property.name}) {
                _result.${property.name} = null;
            } else {
                _result.${property.name} = {};
                ${referenceCode.join('\n')}
            }
            `);
            continue;
        }
        setProperties.push(`
            //createJITConverterForSnapshot ${property.name} class:snapshot:${property.type}
            ${serializer_compiler_1.getDataConverterJS(`_result.${property.name}`, `_value.${property.name}`, property, serializerCompilers, context, jitStack, `_result.${property.name} = null`, `_result.${property.name} = null`)}
            `);
    }
    let circularCheckBeginning = '';
    let circularCheckEnd = '';
    if (schema.hasCircularReference()) {
        circularCheckBeginning = `
        if (_stack) {
            if (_stack.includes(_value)) return undefined;
        } else {
            _stack = [];
        }
        _stack.push(_value);
        `;
        circularCheckEnd = `_stack.pop();`;
    }
    const functionCode = `
        return function(_value, _parents, _options, _stack, _depth) {
            ${circularCheckBeginning}
            var _result = {};
            var oldUnpopulatedCheck = _global.unpopulatedCheck;
            _global.unpopulatedCheck = UnpopulatedCheckNone;
            ${setProperties.join('\n')}
            _global.unpopulatedCheck = oldUnpopulatedCheck;
            ${circularCheckEnd}
            return _result;
        }
        `;
    context.set('_global', model_1.getGlobalStore());
    context.set('UnpopulatedCheckNone', model_1.UnpopulatedCheck.None);
    const compiled = new Function(...context.keys(), functionCode);
    const fn = compiled.bind(undefined, ...context.values())();
    fn.buildId = schema.buildId;
    return fn;
}
/**
 * Creates a new JIT compiled function to convert the class instance to a snapshot.
 * A snapshot is essentially the class instance as `plain` serialization while references are
 * stored only as their primary keys.
 *
 * Generated function is cached.
 */
function getConverterForSnapshot(classSchema) {
    const jit = classSchema.jit;
    if (jit.snapshotConverter)
        return jit.snapshotConverter;
    jit.snapshotConverter = createJITConverterForSnapshot(classSchema, classSchema.getProperties(), json_serializer_1.jsonSerializer.fromClass);
    core_1.toFastProperties(jit);
    return jit.snapshotConverter;
}
exports.getConverterForSnapshot = getConverterForSnapshot;
/**
 * Creates a snapshot using getConverterForSnapshot().
 */
function createSnapshot(classSchema, item) {
    return getConverterForSnapshot(classSchema)(item);
}
exports.createSnapshot = createSnapshot;
/**
 * Extracts the primary key of JSONPartial (snapshot) and converts to class type.
 */
function getPrimaryKeyExtractor(classSchema) {
    const jit = classSchema.jit;
    if (jit.primaryKey)
        return jit.primaryKey;
    jit.primaryKey = createJITConverterForSnapshot(classSchema, classSchema.getPrimaryFields(), json_serializer_1.jsonSerializer.toClass);
    core_1.toFastProperties(jit);
    return jit.primaryKey;
}
exports.getPrimaryKeyExtractor = getPrimaryKeyExtractor;
/**
 * Creates a primary key hash generator that takes an item from any format
 * converts it to class format, then to plain, then uses the primitive values to create a string hash.
 */
function getPrimaryKeyHashGenerator(classSchema, serializer = json_serializer_1.jsonSerializer) {
    const jit = classSchema.jit;
    if (!jit.pkHash) {
        jit.pkHash = {};
        core_1.toFastProperties(jit);
    }
    if (jit.pkHash[serializer.name])
        return jit.pkHash[serializer.name];
    jit.pkHash[serializer.name] = createPrimaryKeyHashGenerator(classSchema, serializer);
    core_1.toFastProperties(jit.pkHash);
    return jit.pkHash[serializer.name];
}
exports.getPrimaryKeyHashGenerator = getPrimaryKeyHashGenerator;
function getForeignKeyHash(row, property) {
    const foreignSchema = property.getResolvedClassSchema();
    return getPrimaryKeyHashGenerator(foreignSchema)(row[property.name]);
}
exports.getForeignKeyHash = getForeignKeyHash;
function simplePrimaryKeyHash(value) {
    return '\0' + value;
}
function getSimplePrimaryKeyHashGenerator(classSchema) {
    return simplePrimaryKeyHash;
}
exports.getSimplePrimaryKeyHashGenerator = getSimplePrimaryKeyHashGenerator;
function createPrimaryKeyHashGenerator(schema, serializer) {
    const context = new Map();
    const setProperties = [];
    const jitStack = new jit_1.JitStack();
    for (const property of schema.getPrimaryFields()) {
        if (property.isParentReference)
            continue;
        if (property.isReference) {
            const referenceCode = [];
            for (const pk of property.getResolvedClassSchema().getPrimaryFields()) {
                if (pk.type === 'class') {
                    throw new Error(`Class as primary key (${property.getResolvedClassSchema().getClassName()}.${pk.name}) is not supported`);
                }
                referenceCode.push(`
                //getPrimaryKeyExtractor ${property.name}->${pk.name} class:snapshot:${property.type} reference
                lastValue = '';
                ${serializer_compiler_1.getDataConverterJS(`lastValue`, `_value.${property.name}.${pk.name}`, pk, serializer.toClass, context, jitStack)}
                ${serializer_compiler_1.getDataConverterJS(`lastValue`, `lastValue`, pk, json_serializer_1.jsonSerializer.fromClass, context, jitStack)}
                _result += '\\0' + lastValue;
            `);
            }
            setProperties.push(`
            //getPrimaryKeyExtractor ${property.name} class:snapshot:${property.type} reference
            if (undefined !== _value.${property.name} && null !== _value.${property.name}) {
                ${referenceCode.join('\n')}
            } else {
                _result += '\\0';
            }
            `);
            continue;
        }
        if (property.type === 'class') {
            throw new Error(`Class as primary key (${schema.getClassName()}.${property.name}) is not supported`);
        }
        setProperties.push(`
            //getPrimaryKeyHashGenerator ${property.name} class:plain:${property.type}
            lastValue = '';
            ${serializer_compiler_1.getDataConverterJS(`lastValue`, `_value.${property.name}`, property, serializer.toClass, context, jitStack)}
            ${serializer_compiler_1.getDataConverterJS(`lastValue`, `lastValue`, property, json_serializer_1.jsonSerializer.fromClass, context, jitStack)}
            _result += '\\0' + lastValue;
        `);
    }
    let circularCheckBeginning = '';
    let circularCheckEnd = '';
    if (schema.hasCircularReference()) {
        circularCheckBeginning = `
        if (_stack) {
            if (_stack.includes(_value)) return undefined;
        } else {
            _stack = [];
        }
        _stack.push(_value);
        `;
        circularCheckEnd = `_stack.pop();`;
    }
    const functionCode = `
        return function(_value, _stack) {
            var _result = '';
            var lastValue;
            ${circularCheckBeginning}
            ${setProperties.join('\n')}
            ${circularCheckEnd}
            return _result;
        }
    `;
    const compiled = new Function(...context.keys(), functionCode);
    const fn = compiled.bind(undefined, ...context.values())();
    fn.buildId = schema.buildId;
    return fn;
}
