// @deepkit/orm/dist/cjs/src/database-adapter.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseAdapter = exports.RawFactory = exports.DatabasePersistence = exports.DatabaseAdapterQueryFactory = void 0;
class DatabaseAdapterQueryFactory {
}
exports.DatabaseAdapterQueryFactory = DatabaseAdapterQueryFactory;
class DatabasePersistence {
}
exports.DatabasePersistence = DatabasePersistence;
class RawFactory {
    create(...args) {
        throw new Error(`Current database adapter does not support raw mode.`);
    }
}
exports.RawFactory = RawFactory;
/**
 * A generic database adapter you can use if the API of `Query` is sufficient.
 *
 * You can specify a more specialized adapter like MysqlDatabaseAdapter/MongoDatabaseAdapter with special API for MySQL/Mongo.
 */
class DatabaseAdapter {
    rawFactory(session) {
        return new RawFactory();
    }
    ;
}
exports.DatabaseAdapter = DatabaseAdapter;
