// @deepkit/mongo/dist/cjs/tests/mongo-query.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
require("reflect-metadata");
const index_1 = require("../index");
const type_1 = require("@deepkit/type");
const mongo_serializer_1 = require("../src/mongo-serializer");
class SimpleConfig {
    constructor(items = []) {
        this.items = [];
        this.items = items;
    }
}
__decorate([
    type_1.t.array(type_1.t.string).decorated,
    __metadata("design:type", Array)
], SimpleConfig.prototype, "items", void 0);
let SimpleConfigRef = class SimpleConfigRef {
    constructor(id) {
        this.id = id;
        this.name = '';
    }
};
__decorate([
    type_1.t,
    __metadata("design:type", String)
], SimpleConfigRef.prototype, "name", void 0);
SimpleConfigRef = __decorate([
    __param(0, type_1.t.primary),
    __metadata("design:paramtypes", [Number])
], SimpleConfigRef);
class Simple {
    constructor() {
        this.config = new SimpleConfig;
        this.tags = [];
    }
}
__decorate([
    type_1.t,
    __metadata("design:type", Number)
], Simple.prototype, "id", void 0);
__decorate([
    type_1.t,
    __metadata("design:type", Number)
], Simple.prototype, "price", void 0);
__decorate([
    type_1.t,
    __metadata("design:type", String)
], Simple.prototype, "label", void 0);
__decorate([
    type_1.t.type(SimpleConfig),
    __metadata("design:type", SimpleConfig)
], Simple.prototype, "config", void 0);
__decorate([
    type_1.t.reference(),
    __metadata("design:type", SimpleConfigRef)
], Simple.prototype, "configRef", void 0);
__decorate([
    type_1.t.array(type_1.t.string),
    __metadata("design:type", Array)
], Simple.prototype, "tags", void 0);
globals_1.test('simple', () => {
    const fieldNames = {};
    const m = index_1.convertPlainQueryToMongo(Simple, {
        id: { $qt: '1' }
    }, fieldNames);
    globals_1.expect(m['id']['$qt']).toBe(1);
    globals_1.expect(Object.keys(fieldNames)).toEqual(['id']);
});
globals_1.test('simple class query', () => {
    const partial = mongo_serializer_1.mongoSerializer.for(Simple).serializeProperty('config', new SimpleConfig(['a', 'b']));
    globals_1.expect(partial).toEqual(['a', 'b']);
    const fieldNames = {};
    const m = index_1.convertClassQueryToMongo(Simple, {
        id: { $qt: '1' },
        config: new SimpleConfig(['a', 'b'])
    }, fieldNames);
    globals_1.expect(m.id['$qt']).toBe(1);
    globals_1.expect(m['config']).toEqual(['a', 'b']);
    globals_1.expect(Object.keys(fieldNames)).toEqual(['id', 'config']);
});
globals_1.test('reference object query', () => {
    const fieldNames = {};
    const m = index_1.convertClassQueryToMongo(Simple, {
        configRef: new SimpleConfigRef(2),
    }, fieldNames);
    globals_1.expect(Object.keys(m)).toEqual(['configRef']);
    globals_1.expect(m['configRef']).toBe(2);
    globals_1.expect(Object.keys(fieldNames)).toEqual(['configRef']);
});
globals_1.test('reference object query $in', () => {
    const fieldNames = {};
    const m = index_1.convertClassQueryToMongo(Simple, {
        configRef: { $in: [new SimpleConfigRef(2)] },
    }, fieldNames);
    globals_1.expect(m['configRef']['$in']).toEqual([2]);
    globals_1.expect(Object.keys(m)).toEqual(['configRef']);
    globals_1.expect(Object.keys(fieldNames)).toEqual(['configRef']);
});
globals_1.test('simple class query array', () => {
    const partial = mongo_serializer_1.mongoSerializer.for(Simple).serializeProperty('config', new SimpleConfig(['a', 'b']));
    globals_1.expect(partial).toEqual(['a', 'b']);
    const fieldNames = {};
    const m = index_1.convertClassQueryToMongo(Simple, {
        $and: [{ id: { $qt: '1' } }],
        $or: [{ id: { $qt: '1' } }],
        $nor: [{ id: { $qt: '1' } }],
        $not: [{ configRef: { $qt: new SimpleConfigRef(2) } }],
    }, fieldNames);
    globals_1.expect(m['$and'][0]['id']['$qt']).toBe(1);
    globals_1.expect(m['$or'][0]['id']['$qt']).toBe(1);
    globals_1.expect(m['$nor'][0]['id']['$qt']).toBe(1);
    globals_1.expect(m['$not'][0]['configRef']['$qt']).toBe(2);
    globals_1.expect(Object.keys(fieldNames)).toEqual(['id', 'configRef']);
});
globals_1.test('convertClassQueryToMongo customMapping', () => {
    {
        const fieldNames = {};
        const m = index_1.convertClassQueryToMongo(Simple, {
            $and: [{ id: { $join: '1,2,3,4' } }],
        }, fieldNames, {
            '$join': (name, value, fieldNamesMap) => {
                return value.split(',').map((v) => Number(v));
            }
        });
        globals_1.expect(fieldNames['id']).toBeUndefined();
        globals_1.expect(m['$and'][0]['id']).toEqual([1, 2, 3, 4]);
    }
    {
        const fieldNames = {};
        const m = index_1.convertClassQueryToMongo(Simple, {
            $and: [{ id: { $join: '1,2,3,4' } }],
        }, fieldNames, {
            '$join': (name, value, fieldNamesMap) => {
                fieldNamesMap[name] = true;
                return value.split(',').map((v) => Number(v));
            }
        });
        globals_1.expect(fieldNames['id']).toBe(true);
    }
    {
        const m = index_1.convertClassQueryToMongo(Simple, {
            id: { $join: '1,2,3,4' },
        }, {}, {
            '$join': (name, value) => {
                return value.split(',').map((v) => Number(v));
            }
        });
        globals_1.expect(m['id']).toEqual([1, 2, 3, 4]);
    }
    {
        const m = index_1.convertClassQueryToMongo(Simple, {
            id: { $join: '1,2,3,4' },
        }, {}, {
            '$join': (name, value) => {
                return undefined;
            }
        });
        globals_1.expect(m['id']).toBeUndefined();
    }
});
globals_1.test('simple 2', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { id: { dif: 1 } });
    globals_1.expect(m).toEqual({ id: NaN });
});
globals_1.test('regex', () => {
    const fieldNames = {};
    const m = index_1.convertPlainQueryToMongo(Simple, { label: { $regex: /0-9+/ } }, fieldNames);
    globals_1.expect(m).toEqual({ label: { $regex: /0-9+/ } });
    globals_1.expect(Object.keys(fieldNames)).toEqual(['label']);
});
globals_1.test('and', () => {
    const fieldNames = {};
    const m = index_1.convertPlainQueryToMongo(Simple, { $and: [{ id: '1' }, { id: '2' }] }, fieldNames);
    globals_1.expect(m).toEqual({ $and: [{ id: 1 }, { id: 2 }] });
    globals_1.expect(m['$and'][0]['id']).toBe(1);
    globals_1.expect(m['$and'][1]['id']).toBe(2);
    globals_1.expect(Object.keys(fieldNames)).toEqual(['id']);
});
globals_1.test('$in', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { id: { $in: ['1', '2'] } });
    globals_1.expect(m).toEqual({ id: { $in: [1, 2] } });
    const m2 = index_1.convertPlainQueryToMongo(Simple, { id: { $nin: ['1', '2'] } });
    globals_1.expect(m2).toEqual({ id: { $nin: [1, 2] } });
});
globals_1.test('implicit array item', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { tags: "one" });
    globals_1.expect(m).toEqual({ tags: "one" });
    const m2 = index_1.convertPlainQueryToMongo(Simple, { tags: 3 });
    globals_1.expect(m2).toEqual({ tags: "3" });
});
globals_1.test('$all', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { tags: { $all: ['one', 'two', 3] } });
    globals_1.expect(m).toEqual({ tags: { $all: ['one', 'two', '3'] } });
    const m2 = index_1.convertPlainQueryToMongo(Simple, { tags: { $not: { $all: ['one', 'two', 3] } } });
    globals_1.expect(m2).toEqual({ tags: { $not: { $all: ['one', 'two', '3'] } } });
    const m3 = index_1.convertPlainQueryToMongo(Simple, { tags: { $not: { $all: [['one', 'two', 3]] } } });
    globals_1.expect(m3).toEqual({ tags: { $not: { $all: [['one', 'two', '3']] } } });
});
globals_1.test('complex', () => {
    const names = {};
    const m = index_1.convertPlainQueryToMongo(Simple, { $and: [{ price: { $ne: '1.99' } }, { price: { $exists: true } }, { id: { $gt: '0' } }] }, names);
    globals_1.expect(m).toEqual({ $and: [{ price: { $ne: 1.99 } }, { price: { $exists: true } }, { id: { $gt: 0 } }] });
    globals_1.expect(Object.keys(names)).toEqual(['price', 'id']);
});
globals_1.test('complex 2', () => {
    const names = {};
    const date = new Date();
    class NodeCluster {
        constructor() {
            this.connected = false;
            this.disabled = false;
        }
    }
    __decorate([
        type_1.t,
        __metadata("design:type", Boolean)
    ], NodeCluster.prototype, "connected", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Boolean)
    ], NodeCluster.prototype, "disabled", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Date)
    ], NodeCluster.prototype, "lastConnectionTry", void 0);
    const m = index_1.convertPlainQueryToMongo(NodeCluster, {
        $and: [
            { connected: false, disabled: { $ne: true } },
            { $or: [{ lastConnectionTry: { $exists: false } }, { lastConnectionTry: { $lt: date } }] }
        ]
    }, names);
    globals_1.expect(m).toEqual({
        $and: [
            { connected: false, disabled: { $ne: true } },
            { $or: [{ lastConnectionTry: { $exists: false } }, { lastConnectionTry: { $lt: date } }] }
        ]
    });
    globals_1.expect(Object.keys(names)).toEqual(['connected', 'disabled', 'lastConnectionTry']);
});
globals_1.test('$or', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { $and: [{ $or: [{ id: 1 }] }] });
    globals_1.expect(m).toEqual({ $and: [{ $or: [{ id: 1 }] }] });
});
globals_1.test('nested $or', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { $or: [{ id: { $lt: 20 } }, { price: 10 }] });
    globals_1.expect(m).toEqual({ $or: [{ id: { $lt: 20 } }, { price: 10 }] });
});
globals_1.test('not', () => {
    const m = index_1.convertPlainQueryToMongo(Simple, { $not: [{ price: { $ne: '1.99' } }, { price: { $exists: true } }] });
    globals_1.expect(m).toEqual({ $not: [{ price: { $ne: 1.99 } }, { price: { $exists: true } }] });
});
globals_1.test('excluded', () => {
    const excluded = ['$exists', '$mod', '$size', '$type', '$regex', '$where'];
    for (const e of excluded) {
        const obj = { label: {} };
        obj['label'][e] = true;
        const m = index_1.convertPlainQueryToMongo(Simple, obj);
        globals_1.expect(m).toEqual(obj);
    }
});
