// @deepkit/sql/dist/cjs/tests/inheritance.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const type_1 = require("@deepkit/type");
const schema_parser_1 = require("../src/reverse/schema-parser");
const default_platform_1 = require("../src/platform/default-platform");
const globals_1 = require("@jest/globals");
let Person = class Person {
    constructor() {
        this.id = 0;
    }
};
__decorate([
    type_1.t.primary.autoIncrement,
    __metadata("design:type", Number)
], Person.prototype, "id", void 0);
__decorate([
    type_1.t,
    __metadata("design:type", String)
], Person.prototype, "firstName", void 0);
__decorate([
    type_1.t,
    __metadata("design:type", String)
], Person.prototype, "lastName", void 0);
__decorate([
    type_1.t,
    __metadata("design:type", String)
], Person.prototype, "type", void 0);
Person = __decorate([
    type_1.entity.name('person').collectionName('persons')
], Person);
let Employee = class Employee extends Person {
    constructor() {
        super(...arguments);
        this.type = 'employee';
    }
};
__decorate([
    type_1.t,
    __metadata("design:type", String)
], Employee.prototype, "email", void 0);
__decorate([
    type_1.t.literal('employee'),
    __metadata("design:type", String)
], Employee.prototype, "type", void 0);
Employee = __decorate([
    type_1.entity.name('employee').singleTableInheritance()
], Employee);
let Freelance = class Freelance extends Person {
    constructor() {
        super(...arguments);
        this.type = 'freelancer';
    }
};
__decorate([
    type_1.t,
    __metadata("design:type", String)
], Freelance.prototype, "token", void 0);
__decorate([
    type_1.t.literal('freelancer'),
    __metadata("design:type", String)
], Freelance.prototype, "type", void 0);
Freelance = __decorate([
    type_1.entity.name('freelancer').singleTableInheritance()
], Freelance);
class MySchemaParser extends schema_parser_1.SchemaParser {
    parse(database, limitTableNames) {
    }
}
class MyPlatform extends default_platform_1.DefaultPlatform {
    constructor() {
        super();
        this.schemaParserType = MySchemaParser;
        this.addType('number', 'integer');
    }
}
globals_1.test('tables', () => {
    const platform = new MyPlatform();
    const tables = platform.createTables([Employee, Freelance]);
    globals_1.expect(tables.length).toBe(1);
    const table = tables[0];
    globals_1.expect(table.columns.length).toBe(6);
    globals_1.expect(table.getColumn('type').type).toBe('text');
    globals_1.expect(table.getColumn('type').isNotNull).toBe(true);
});
