// @deepkit/sql/dist/cjs/tests/sql-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@jest/globals");
const type_1 = require("@deepkit/type");
const sql_filter_builder_1 = require("../src/sql-filter-builder");
const sqlstring_1 = require("sqlstring");
function quoteId(value) {
    return value;
}
globals_1.test('QueryToSql', () => {
    const User = type_1.t.class({
        id: type_1.t.number,
        username: type_1.t.string,
        password: type_1.t.string,
        disabled: type_1.t.boolean,
        created: type_1.t.date,
    });
    const queryToSql = new sql_filter_builder_1.SQLFilterBuilder(type_1.getClassSchema(User), quoteId('user'), type_1.jsonSerializer, sqlstring_1.escape, quoteId);
    globals_1.expect(queryToSql.convert({ id: 123 })).toBe(`user.id = 123`);
    globals_1.expect(queryToSql.convert({ id: '$id' })).toBe(`user.id = user.id`);
    globals_1.expect(queryToSql.convert({ username: 'Peter' })).toBe(`user.username = 'Peter'`);
    globals_1.expect(queryToSql.convert({ id: 44, username: 'Peter' })).toBe(`(user.id = 44 AND user.username = 'Peter')`);
    globals_1.expect(queryToSql.convert({ $or: [{ id: 44 }, { username: 'Peter' }] })).toBe(`(user.id = 44 OR user.username = 'Peter')`);
    globals_1.expect(queryToSql.convert({ $and: [{ id: 44 }, { username: 'Peter' }] })).toBe(`(user.id = 44 AND user.username = 'Peter')`);
    globals_1.expect(queryToSql.convert({ id: { $ne: 44 } })).toBe(`user.id != 44`);
    globals_1.expect(queryToSql.convert({ id: { $eq: 44 } })).toBe(`user.id = 44`);
    globals_1.expect(queryToSql.convert({ id: { $gt: 44 } })).toBe(`user.id > 44`);
    globals_1.expect(queryToSql.convert({ id: { $gte: 44 } })).toBe(`user.id >= 44`);
    globals_1.expect(queryToSql.convert({ id: { $lt: 44 } })).toBe(`user.id < 44`);
    globals_1.expect(queryToSql.convert({ id: { $lte: 44 } })).toBe(`user.id <= 44`);
    globals_1.expect(queryToSql.convert({ id: { $in: [44, 55] } })).toBe(`user.id IN (44,55)`);
    globals_1.expect(queryToSql.convert({ id: { $nin: [44, 55] } })).toBe(`user.id NOT IN (44,55)`);
    globals_1.expect(() => queryToSql.convert({ id: { $oasdads: 123 } })).toThrow('not supported');
});
