// @deepkit/mongo/dist/cjs/src/client/command/auth/x509.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.X509Auth = void 0;
const command_1 = require("../command");
const type_1 = require("@deepkit/type");
class AuthenticateCommand extends type_1.t.class({
    authenticate: type_1.t.literal(1),
    mechanism: type_1.t.string,
    $db: type_1.t.string,
    username: type_1.t.string.optional,
}) {
}
class AuthenticateResponse extends type_1.t.extendClass(command_1.BaseResponse, {}) {
}
class X509Auth {
    async auth(command, config) {
        await command.sendAndWait(AuthenticateCommand, {
            authenticate: 1,
            mechanism: 'MONGODB-X509',
            $db: '$external',
            username: config.authUser
        }, AuthenticateResponse);
    }
}
exports.X509Auth = X509Auth;
