// @deepkit/rpc/dist/cjs/src/decorators.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getActions = exports.getActionParameters = exports.rpc = exports.rpcProperty = exports.rpcClass = exports.RpcAction = void 0;
const type_1 = require("@deepkit/type");
class RpcController {
    constructor() {
        this.actions = new Map();
    }
    getPath() {
        return this.definition ? this.definition.path : this.name || '';
    }
}
class RpcAction {
    constructor() {
        this.groups = [];
    }
}
exports.RpcAction = RpcAction;
class RpcClass {
    constructor() {
        this.t = new RpcController;
    }
    controller(nameOrDefinition) {
        if ('string' === typeof nameOrDefinition) {
            this.t.name = nameOrDefinition;
        }
        else {
            this.t.definition = nameOrDefinition;
        }
    }
    addAction(name, action) {
        this.t.actions.set(name, action);
    }
}
exports.rpcClass = type_1.createClassDecoratorContext(RpcClass);
class RpcProperty {
    constructor() {
        this.t = new RpcAction;
    }
    onDecorator(classType, property) {
        this.t.name = property;
        this.t.classType = classType;
        exports.rpcClass.addAction(property, this.t)(classType);
    }
    action() {
    }
    group(...groups) {
        this.t.groups.push(...groups);
    }
}
exports.rpcProperty = type_1.createPropertyDecoratorContext(RpcProperty);
exports.rpc = type_1.mergeDecorator(exports.rpcClass, exports.rpcProperty);
function getActionParameters(target, method) {
    return type_1.getClassSchema(target).getMethodProperties(method);
}
exports.getActionParameters = getActionParameters;
function getActions(target) {
    var _a, _b;
    return (_b = (_a = exports.rpcClass._fetch(target)) === null || _a === void 0 ? void 0 : _a.actions) !== null && _b !== void 0 ? _b : new Map;
}
exports.getActions = getActions;
