// @deepkit/framework-debug-api/dist/esm/src/api.js
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { entity, PropertySchema, t } from '@deepkit/type';
import { ControllerSymbol } from '@deepkit/rpc';
import { DebugRequest } from './model';
export class ConfigOption {
}
__decorate([
    t,
    __metadata("design:type", String)
], ConfigOption.prototype, "name", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], ConfigOption.prototype, "type", void 0);
__decorate([
    t.any,
    __metadata("design:type", Object)
], ConfigOption.prototype, "defaultValue", void 0);
__decorate([
    t.any,
    __metadata("design:type", Object)
], ConfigOption.prototype, "value", void 0);
__decorate([
    t.optional,
    __metadata("design:type", String)
], ConfigOption.prototype, "description", void 0);
let Workflow = class Workflow {
};
__decorate([
    t.array(t.string),
    __metadata("design:type", Array)
], Workflow.prototype, "places", void 0);
__decorate([
    t.array(t.schema({ from: t.string, to: t.string, label: t.string.optional })),
    __metadata("design:type", Array)
], Workflow.prototype, "transitions", void 0);
Workflow = __decorate([
    entity.name('debug/workflow')
], Workflow);
export { Workflow };
let DatabaseEntity = class DatabaseEntity {
};
__decorate([
    t.optional,
    __metadata("design:type", String)
], DatabaseEntity.prototype, "name", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], DatabaseEntity.prototype, "className", void 0);
DatabaseEntity = __decorate([
    entity.name('debug/database/entity')
], DatabaseEntity);
export { DatabaseEntity };
let Database = class Database {
    constructor() {
        this.entities = [];
    }
};
__decorate([
    t,
    __metadata("design:type", String)
], Database.prototype, "name", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], Database.prototype, "adapter", void 0);
__decorate([
    t.array(DatabaseEntity),
    __metadata("design:type", Array)
], Database.prototype, "entities", void 0);
Database = __decorate([
    entity.name('debug/database')
], Database);
export { Database };
let Config = class Config {
};
__decorate([
    t.array(ConfigOption),
    __metadata("design:type", Array)
], Config.prototype, "appConfig", void 0);
__decorate([
    t.array(ConfigOption),
    __metadata("design:type", Array)
], Config.prototype, "modulesConfig", void 0);
Config = __decorate([
    entity.name('debug/config')
], Config);
export { Config };
export class RouteParameter {
}
__decorate([
    t,
    __metadata("design:type", String)
], RouteParameter.prototype, "name", void 0);
__decorate([
    t.string,
    __metadata("design:type", String)
], RouteParameter.prototype, "type", void 0);
__decorate([
    t.any,
    __metadata("design:type", Object)
], RouteParameter.prototype, "schema", void 0);
let Route = class Route {
    constructor(path, httpMethod, controller, description, parameters, groups, category, bodySchema) {
        this.path = path;
        this.httpMethod = httpMethod;
        this.controller = controller;
        this.description = description;
        this.parameters = parameters;
        this.groups = groups;
        this.category = category;
        this.bodySchema = bodySchema;
        if (bodySchema) {
            if (bodySchema.classType) {
                //we don't and can't instantiate the full PropertySchema, since the
                //type is not available at runtime.
                bodySchema.classTypeName = bodySchema.classType;
                bodySchema.classType = undefined;
            }
            this.bodyPropertySchema = PropertySchema.fromJSON(bodySchema);
        }
    }
};
Route = __decorate([
    entity.name('debug/route'),
    __param(0, t.name('path')),
    __param(1, t.name('httpMethod')),
    __param(2, t.name('controller')),
    __param(3, t.name('description')),
    __param(4, t.array(RouteParameter).name('parameters')),
    __param(5, t.array(t.string).name('groups')),
    __param(6, t.string.name('category')),
    __param(7, t.any.name('bodySchema')),
    __metadata("design:paramtypes", [String, String, String, String, Array, Array, String, Object])
], Route);
export { Route };
let RpcActionParameter = class RpcActionParameter {
    constructor(name, schema) {
        this.name = name;
        this.schema = schema;
        this.propertySchema = PropertySchema.fromJSON(schema);
    }
};
RpcActionParameter = __decorate([
    entity.name('rpc/action/parameter'),
    __param(0, t.name('name')),
    __param(1, t.any.name('schema')),
    __metadata("design:paramtypes", [String, Object])
], RpcActionParameter);
export { RpcActionParameter };
let RpcAction = class RpcAction {
};
__decorate([
    t,
    __metadata("design:type", String)
], RpcAction.prototype, "path", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], RpcAction.prototype, "controller", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], RpcAction.prototype, "methodName", void 0);
__decorate([
    t.array(RpcActionParameter),
    __metadata("design:type", Array)
], RpcAction.prototype, "parameters", void 0);
RpcAction = __decorate([
    entity.name('rpc/action')
], RpcAction);
export { RpcAction };
let Event = class Event {
};
__decorate([
    t,
    __metadata("design:type", String)
], Event.prototype, "event", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], Event.prototype, "controller", void 0);
__decorate([
    t,
    __metadata("design:type", String)
], Event.prototype, "methodName", void 0);
__decorate([
    t,
    __metadata("design:type", Number)
], Event.prototype, "priority", void 0);
Event = __decorate([
    entity.name('rpc/event')
], Event);
export { Event };
export const DebugControllerInterface = ControllerSymbol('deepkit/debug/controller', [Config, Database, Route, RpcAction, Workflow, Event, DebugRequest]);
