// @deepkit/http/dist/cjs/src/model.js
"use strict";
/*
 * Deepkit Framework
 * Copyright (C) 2021 Deepkit UG, Marc J. Schmidt
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the MIT License.
 *
 * You should have received a copy of the MIT License along with this program.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpRequest = exports.RequestBuilder = exports.HttpResponse = void 0;
const http_1 = require("http");
const net_1 = require("net");
const querystring = __importStar(require("querystring"));
class HttpResponse extends http_1.ServerResponse {
    status(code) {
        this.writeHead(code);
        this.end();
    }
}
exports.HttpResponse = HttpResponse;
class RequestBuilder {
    constructor(path, method = 'GET') {
        this.path = path;
        this.method = method;
        this.contentBuffer = Buffer.alloc(0);
        this.headers = {};
    }
    getUrl() {
        if (this.queryPath) {
            return this.path + '?' + this.queryPath;
        }
        return this.path;
    }
    build() {
        const headers = this.headers;
        const method = this.method;
        const url = this.getUrl();
        const bodyContent = this.contentBuffer;
        return new (class extends HttpRequest {
            constructor() {
                super(...arguments);
                this.url = url;
                this.method = method;
                this.position = 0;
                this.headers = headers;
                this.done = false;
            }
            _read(size) {
                if (this.done) {
                    this.push(null);
                }
                else {
                    this.push(bodyContent);
                    this.done = true;
                }
            }
        })(new net_1.Socket());
    }
    header(name, value) {
        this.headers[name] = String(value);
        return this;
    }
    json(body) {
        this.contentBuffer = Buffer.from(JSON.stringify(body), 'utf8');
        this.headers['Content-Type'] = 'application/json';
        this.headers['Content-Length'] = String(this.contentBuffer.byteLength);
        return this;
    }
    body(body) {
        if ('string' === typeof body) {
            this.contentBuffer = Buffer.from(body, 'utf8');
        }
        else {
            this.contentBuffer = body;
        }
        this.headers['Content-Length'] = String(this.contentBuffer.byteLength);
        return this;
    }
    query(query) {
        this.queryPath = querystring.stringify(query);
        return this;
    }
}
exports.RequestBuilder = RequestBuilder;
class HttpRequest extends http_1.IncomingMessage {
    constructor() {
        super(...arguments);
        /**
         * A store that can be used to transport data from guards/listeners to ParameterResolvers/controllers.
         */
        this.store = {};
        this.uploadedFiles = {};
    }
    static GET(path) {
        return new RequestBuilder(path);
    }
    static POST(path) {
        return new RequestBuilder(path, 'POST');
    }
    static PUT(path) {
        return new RequestBuilder(path, 'PUT');
    }
    static DELETE(path) {
        return new RequestBuilder(path, 'DELETE');
    }
    getUrl() {
        return this.url || '/';
    }
    getMethod() {
        return this.method || 'GET';
    }
    getRemoteAddress() {
        return this.connection.remoteAddress || '';
    }
}
exports.HttpRequest = HttpRequest;
