// @deepkit/bson/dist/cjs/tests/case/partial-value.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const type_1 = require("@deepkit/type");
const globals_1 = require("@jest/globals");
require("reflect-metadata");
const bson_jit_parser_1 = require("../../src/bson-jit-parser");
const bson_serialize_1 = require("../../src/bson-serialize");
globals_1.test('partial', () => {
    class User {
        constructor() {
            this.id = 0;
            this.name = '';
        }
    }
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], User.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", String)
    ], User.prototype, "name", void 0);
    class Model {
        constructor() {
            this.id = 0;
            this.defaultValue = 5;
        }
    }
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Model.prototype, "id", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", Number)
    ], Model.prototype, "defaultValue", void 0);
    __decorate([
        type_1.t,
        __metadata("design:type", User)
    ], Model.prototype, "manager", void 0);
    {
        const s = type_1.t.schema({
            v: type_1.t.type(Model),
        });
        const bson = bson_serialize_1.getBSONSerializer(s)({ v: { id: 4 } });
        const value = bson_jit_parser_1.getBSONDecoder(s)(bson);
        globals_1.expect(value.v).toBeInstanceOf(Model);
        globals_1.expect(value).toEqual({ v: { id: 4, defaultValue: 5 } });
    }
    {
        const s = type_1.t.schema({
            v: type_1.t.partial(Model),
        });
        {
            const bson = bson_serialize_1.getBSONSerializer(s)({ v: { id: 4 } });
            const value = bson_jit_parser_1.getBSONDecoder(s)(bson);
            globals_1.expect(value.v).not.toBeInstanceOf(Model);
            globals_1.expect(value).toEqual({ v: { id: 4 } });
        }
        {
            const bson = bson_serialize_1.getBSONSerializer(s)({ v: { id: 4, manager: { id: 23, name: 'peter' } } });
            const value = bson_jit_parser_1.getBSONDecoder(s)(bson);
            globals_1.expect(value.v).not.toBeInstanceOf(Model);
            globals_1.expect(value.v.manager).toBeInstanceOf(User); //in partials we have full instances
            globals_1.expect(value).toEqual({ v: { id: 4, manager: { id: 23, name: 'peter' } } });
        }
    }
    {
        const s = type_1.t.schema({
            v: type_1.t.array(type_1.t.partial(Model)),
        });
        {
            const bson = bson_serialize_1.getBSONSerializer(s)({ v: [{ id: 4 }, { id: 5 }] });
            const value = bson_jit_parser_1.getBSONDecoder(s)(bson);
            globals_1.expect(value.v).not.toBeInstanceOf(Model);
            globals_1.expect(value).toEqual({ v: [{ id: 4 }, { id: 5 }] });
        }
    }
});
